<html>
   <head>
      <title>BioThings Explorer</title>
      <meta name = "viewport" content = "width = device-width, initial-scale = 1">      
      <link rel = "stylesheet"
         href = "https://fonts.googleapis.com/icon?family=Material+Icons">
      <link rel = "stylesheet"
         href = "https://cdnjs.cloudflare.com/ajax/libs/materialize/0.97.3/css/materialize.min.css">
      <link href="./static/css/tutorial_style.css" rel="stylesheet" />
      <script type = "text/javascript"
         src = "https://code.jquery.com/jquery-2.1.1.min.js"></script>
      <script src = "https://cdnjs.cloudflare.com/ajax/libs/materialize/0.97.3/js/materialize.min.js">
      </script>
   </head>
   <body> 
      <!-- This section is for the header-->
      <div class="row">
         <div class="col s12" id="navbar">
            <nav>
                <div class="nav-wrapper orange">
                    <a href="/explorer" class="brand-logo">BioThings Explorer</a>
                    <ul class="right hide-on-med-and-down">
                    <li><a href="#" id="map">TUTORIAL</a></li>
                    <li><a href="/explorer" id="explore">EXPLORE</a></li>
                    </ul>
                </div>
            </nav>
         </div>
      </div>
      <!-- This section is Introduction about BioThings Explorer-->
      <div class="container">
         <div class="col s12" id="introduction">
            <h3> What is BioThings Explorer? </h3>
            <p> Current biological research relies significantly on the integration and cross-linking of data across multiple domains, including genes, drugs and diseases. Meanwhile, lots of data providers start to use web-based APIs as an alternative data access point. BioThings Explorer integrate different biological data together through a distributed approach, which is to interlink those individually maintained APIs for integrated knowledge exploration.</p>
            <p>BioThings Explorer utilize the <a href="http://smart-api.info/" target="_blank">SmartAPI</a> basd API-specific metadata and <a href="https://json-ld.org/" target="_blank">JSON-LD</a> based semantic context to solve the API data interoperability issue. By utilizing all these technologies, we are able to build a highly connected network of biological APIs, which are discoverable, reusable and actionable.  More importantly, it empowers users to perform knowledge exploration tasks by identifying potential connecting paths and chaining multiple API calls together.</p>
            <p> You may check <a href="#sources">here</a> for API resources currently integrated in BioThings Explorer</p>
         </div>
         <br>
         <div class="col s12" id="function">
            <h3> What can BioThings Explorer Offer? </h3>
            <br>
            <div class="row">
               <div class="col s3">
                  <div class="center">
                     <a href="#overview">
                       <i class="large material-icons">map</i>
                       <h5 class="promo-caption">Map</h5></a>
                     <p class="light center">Overview of how API, endpoints and bio-entities are connected</p>
                  </div>
               </div>
               <div class="col s3">
                  <div class="center">
                     <a href="#metadata">
                       <i class="large material-icons">search</i>
                       <h5 class="promo-caption">MetaData</h5></a>
                     <p class="light center">Get metadata information about API endpoints</p>
                  </div>
               </div>
               <div class="col s3">
                  <div class="center">
                     <a href="#path">
                       <i class="large material-icons">timeline</i>
                       <h5 class="promo-caption">Path</h5></a>
                     <p class="light center">Find Connection between Bio-entities</p>
                  </div>
               </div>
               <div class="col s3">
                  <div class="center">
                     <a href="#knowledge_explore">
                       <i class="large material-icons">explore</i>
                       <h5 class="promo-caption">Explore</h5></a>
                     <p class="light center">Knowledge Exploration through chaining API calls </p>
                  </div>
               </div>
            </div>
            <br>
            <div class="row" id="overview">
               <div class="col s12 center">
                  <i class="Large material-icons blue-text text-lighten-1">map</i> <h3 class="blue-text text-lighten-1"> Map </h3>
               </div>
               <div class="col s12">
                  <p> BioThings Explorer takes metadata information from SmartAPI specifications, and build a highly connected network of APIs. Users may have an idea of how APIs as well as biological knowledge are connected through three different view point offered by BioThings Explorer.</p>
                  <p> The first view point, also the default one, is the semantic type view. In this scenario, all bio-entities are categorized by their semantic type, e.g. gene, variant, disease, etc. This is a high-level view of the biological information as well as their relationships contained in different APIs. </p>
                  <div class="center">
                     <img src="../static/img/tutorial_semantic_view.png" alt="Select Endpoint View" width="80%">
                  </div>
                  <br>
                  <p> The second view point is the biological ID view. In this scenario, we show the relationship between all biological entities enclosed in our network </p>
                  <div class="center">
                     <img src="../static/img/id_view.png" alt="Select Endpoint View" width="80%">
                  </div>
                  <br>
                  <p> Users may enlarge the graph to view the detailed relationship information between different biological entities. </p>
                  <div class="center">
                     <img src="../static/img/id_view_enlarge.png" alt="Select Endpoint View" width="80%">
                  </div>
                  <br>
                  <p> Users may also hover your mouse on any of the edge to view which API endpoint connects these two biological entities. </p>
                  <div class="center">
                     <img src="../static/img/id_view_hover.png" alt="Select Endpoint View" width="80%">
                  </div>
                  <p> The third view point is the API level view. In this scenario, we cluster biological IDs together based on the APIs which use these IDs as inputs or outputs </p>
                  <div class="center">
                     <img src="../static/img/api_view.png" alt="Select Endpoint View" width="80%">
                  </div>
                  <br>
                  <p> Users may enlarge the graph to view the detailed information about the a specific API. </p>
                  <div class="center">
                     <img src="../static/img/api_view_enlarge.png" alt="Select Endpoint View" width="80%">
                  </div>
              </div>
              <br>
              <div class="row" id="metadata">
                <div class="col s12 center">
                  <i class="Large material-icons blue-text text-lighten-1">search</i> <h3 class="blue-text text-lighten-1"> Metadata </h3>
                 </div>
                 <div class="col s12">
                    <p> BioThings Explorer feeds on OpenAPI specifications stored at <a href="http://smart-api.info/" target="_blank">SmartAPI.info</a>, which includes metadata information of the input and output of individual API endpoint. </p>
                    <p> Users may select any API endpoint of their interest at 'Explore individual Endpoint'. </p>
                    <div class="center">
                       <img src="../static/img/tutorial_select_endpoint.png" alt="Select Endpoint View" width="200" height="300">
                    </div>
                    <p>Once an API endpoint is selected, a Sankey plot would be displayed on the right side of the window to show the input and ouptut information of the API endpoint</p>
                    <div class="center">
                       <img src="../static/img/tutorial_endpoint_sankey.png" alt="Select Endpoint View" width="80%">
                    </div>
                    <p> On the left side of the Sankey plot is all the inputs accepted by the chosen API endpoint; and on the right side of the Sankey plot is all the possible outputs provided by the chosen API endpoint.</p>
                    <br>
                 </div>
              </div>
              <br>
              <div class="row" id="path">
                 <div class="col s12 center">
                    <i class="Large material-icons blue-text text-lighten-1">timeline</i> <h3 class="blue-text text-lighten-1"> Path </h3>
                 </div>
                 <p> Users may explore which API endpoints could connect from or to a specific biological entity, e.g. ncbigene at 'Explore individual Biological Entity as Input' or 'Explore Individual Biological Entity as Ouptut. </p>
                 <p> You may first select any biological entity of your interest</p>
                 <div class="center">
                    <img src="../static/img/tutorial_select_bioentity.png" alt="Select Endpoint View" width="200" height="300">
                 </div>
                 <p>Once an API endpoint is selected, a Sankey plot would be displayed on the right side of the window to show which API endpoint(s) could be connected from or to this bio-entity. The example below shows all API endpoints which could be connected from Uniprot.</p>
                 <div class="center">
                    <img src="../static/img/tutorial_bioentity_sankey.png" alt="Select Bio-Entity View" width="80%">
                 </div>
              </div>
              <br>
              <div class="row" id="knowledge_explore">
                 <div class="col s12 center">
                    <i class="Large material-icons blue-text text-lighten-1">explore</i> <h3 class="blue-text text-lighten-1"> Explore </h3>
                 </div>
                 <p>This part is designed to help users uncover implicit relationship between two biological IDs. Graph theory algorithms are used here to help users identify potential paths (e.g. API endpoints) connecting two biological IDs.</p>
                 <p> Users may first specify the input ID and output ID they would like to explore.</p>
                 <div class="center">
                    <img src="../static/img/tutorial_path_input_output.png" alt="Select Endpoint View" width="200" height="300">
                 </div>
                 <p> Users may also specify the maximum number of APIs used in the path. By default, it is set to be 3, which means at most 3 APIs will be used to construct one path connecting from input to output</p>
                 <div class="center">
                    <img src="../static/img/tutorial_max_api.png" alt="Select Endpoint View" width="200" height="300">
                 </div>
                 <p> Once you finish chosing input, output and max_api, you may click 'FIND PATH' button below. And a Sankey plot would be constructed for you on the right side of the window, showing all paths which could be connected from your input to your output. 
                 <div class="center">
                    <img src="../static/img/tutorial_path_sankey.png" alt="Select Bio-Entity View" width="80%">
                 </div>
                 <p> The example above shows all paths connecting from ncbigene ID to drugname.</p>
                 <p> Below the Sankey plot, there will also be another section which list all these paths</p>
                 <div class="center">
                    <img src="../static/img/tutorial_path_list.png" alt="Select Bio-Entity View" width="80%">
                 </div>
                 <p> Users may then chose any path they desire on the left side of the window, and specify the input value.</p>
                 <div class="center">
                    <img src="../static/img/tutorial_path_select.png" alt="Select Endpoint View" width="200" height="300">
                 </div>
                 <p> Here we select path 11, which is the path connecting from ncbigene -> mygene.info -> hgnc.symbol -> dgidb -> drugname. The input value we specified here is 1018, which correspond to ncbigene ID 1018. You may then click 'EXPLORE THIS PATH', and API calls will be chained together and automatically generated for you. A cytoscape network graph would be displayed on the right, showing all intermediate results as well as final results.</p>
                 <div class="center">
                    <img src="../static/img/tutorial_explore.png" alt="Select Bio-Entity View" width="80%">
                 </div>
                 <p> The graph above shows that the exploration starts from ncbigene:1018 (red dot), and it goes to hgnc.symbol:CDK3 first, and then it connects to three drugnames, which are RGB-286638, ALVOCIDIB and TG-02. The relationship between these nodes are labeled in the edges of the graph.</p>
              </div>
           </div>
           <br>
         <div class="col s12" id="sources">
            <h3> What API resources has been integrated in BioThings Explorer </h3>
            <br>
           <ul class="collection">
             <li class="collection-item avatar">
               <img src="../static/img/mygene_info_logo.svg" alt="" class="circle">
               <span class="title">MyGene.info API</span>
               <p> <br>MyGene.info aggregates gene annotation information from NCBI, Ensembl, Uniprot, NetAffy, PharmGKB, UCSC, CPDB, etc.
               </p>
               <a href="http://mygene.info/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/myvariant_info_logo_nomask.svg" alt="" class="circle">
               <span class="title">MyVariant.info API</span>
               <p> <br>MyVariant.info aggregates variant annotation information from dbSNP, dbNSFP, ClinVar, CADD, COSMIC, UniProt, CIViC, CGI, etc.
               </p>
               <a href="http://myvariant.info/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/mychem_on.png" alt="" class="circle">
               <span class="title">MyChem.info API</span>
               <p> <br>MyChem.info aggregates variant annotation information from ChEBI, ChEMBL, DrugBank, Drug Central, PharmGKB, PubChem, UNII, etc.
               </p>
               <a href="http://mychem.info/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/monarch_logo.png" alt="" class="circle">
               <span class="title">BioLink API</span>
               <p> <br>BioLink API aggregates biological information from variety of databases, covering gene, disease, pathway, genotype, investigations, etc.
               </p>
               <a href="https://api.monarchinitiative.org/api/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/dgidb_logo.png" alt="" class="circle">
               <span class="title">DGIdb API</span>
               <p> <br>DGIdb offer information on drug-gene interactionsfrom over thirty trusted sources, including Ensembl, Entrez, CGI, CIViC, FDA, MyCancerGenome, NCI, etc.
               </p>
               <a href="http://www.dgidb.org/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/do_logo.png" alt="" class="circle">
               <span class="title">Disease Ontology API</span>
               <p> <br>The Disease Ontology ntegrates disease and medical vocabularies through mapping of DO terms to MeSH, ICD, NCIâ€™s thesaurus, SNOMED and OMIM.
               </p>
               <a href="http://disease-ontology.org/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/ols-logo.png" alt="" class="circle">
               <span class="title">Ontology Lookup Service API</span>
               <p> <br>The Ontology Lookup Service (OLS) is a repository for biomedical ontologies that aims to provide a single point of access to the latest ontology versions.
               </p>
               <a href="https://www.ebi.ac.uk/ols/index" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/ensembl_logo.png" alt="" class="circle">
               <span class="title">Ensembl API</span>
               <p> <br>The Ensembl API covers a variety of biological concepts, including variants, transcripts, sequences, phenotypes, regulation, taxonomy, ontology, etc.
               </p>
               <a href="https://rest.ensembl.org/" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/broad_logo.png" alt="" class="circle">
               <span class="title">Broad PGM API</span>
               <p> <br>The Broad-PGM API provides probabilistic graphical models for gene expression data.
               </p>
               <a href="https://github.com/mjstealey/broad-pgm-translator" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
             <li class="collection-item avatar">
               <img src="../static/img/human_base_logo.png" alt="" class="circle">
               <span class="title">HumanBase API</span>
               <p> <br>The HumanBase API provides data-driven predictions of gene expression, function, regulation and interactions in human
               </p>
               <a href="https://github.com/mjstealey/broad-pgm-translator" target="_blank" class="secondary-content"><i class="material-icons">link</i></a>
             </li>
           </ul>
   </body>
</html>