{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biothings_explorer import BioThingsExplorer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bt_explorer = BioThingsExplorer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "registry = bt_explorer.registry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'http://mygene.info/v3/querygene': {'get': {'summary': 'Make gene query and return matching gene hits',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"CDK2\", \"NM_052827\", \"204639_at\". The detailed query syntax can be found at http://docs.mygene.info/en/latest/doc/query_service.html',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/hgnc.symbol/',\n",
       "      'http://identifiers.org/refseq/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hgnc.symbol/',\n",
       "       'template': '/query?q=symbol:{geneid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/refseq/',\n",
       "       'template': '/query?q=refseq.genomic:{geneid}&size=100'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': 'hits.entrezgene',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'hits.symbol',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'mygene.info/jsonld_context/mygene_query_object.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mygene.info/jsonld_context/mygene_query_object.json',\n",
       "  'output': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/ncbigene/': {'assoc:ThingToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ThingToGeneAssociation'],\n",
       "  'input': ['http://identifiers.org/hgnc.symbol/',\n",
       "   'http://identifiers.org/refseq/'],\n",
       "  'api': 'MyGene.info API',\n",
       "  'server': 'http://mygene.info/v3'},\n",
       " 'http://mygene.info/v3/querygeneproduct': {'get': {'summary': 'Make gene query and return matching gene hits',\n",
       "   'parameters': [{'name': 'geneproductid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"CDK2\", \"NM_052827\", \"204639_at\". The detailed query syntax can be found at http://docs.mygene.info/en/latest/doc/query_service.html',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/unigene/',\n",
       "      'http://identifiers.org/uniprot/',\n",
       "      'http://identifiers.org/pdb/',\n",
       "      'http://biothings.io/explorer/vocab/terms/ensembl.protein/',\n",
       "      'http://biothings.io/explorer/vocab/terms/ensembl.transcript/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/unigene/',\n",
       "       'template': '/query?q=unigene:{geneproductid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/uniprot/',\n",
       "       'template': '/query?q=uniprot.Swiss-Prot:{geneproductid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/pdb/',\n",
       "       'template': '/query?q=pdb:{geneproductid}&size=100'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.protein/',\n",
       "       'template': '/query?q=ensembl.protein:{geneproductid}&size=100'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.transcript/',\n",
       "       'template': '/query?q=ensembl.transcript:{geneproductid}&size=100'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': 'hits.entrezgene',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'hits.symbol',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'mygene.info/jsonld_context/mygene_query_object.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mygene.info/jsonld_context/mygene_query_object.json',\n",
       "  'output': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/ncbigene/': {'assoc:ThingToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ThingToGeneAssociation'],\n",
       "  'input': ['http://identifiers.org/unigene/',\n",
       "   'http://identifiers.org/uniprot/',\n",
       "   'http://identifiers.org/pdb/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.protein/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.transcript/'],\n",
       "  'api': 'MyGene.info API',\n",
       "  'server': 'http://mygene.info/v3'},\n",
       " 'http://mygene.info/v3/querypathway': {'get': {'summary': 'Make gene query and return matching gene hits',\n",
       "   'parameters': [{'name': 'pathwayid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"CDK2\", \"NM_052827\", \"204639_at\". The detailed query syntax can be found at http://docs.mygene.info/en/latest/doc/query_service.html',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/biocarta.pathway/',\n",
       "      'http://identifiers.org/kegg.pathway/',\n",
       "      'http://identifiers.org/wikipathways/',\n",
       "      'http://identifiers.org/pharmgkb.pathways/',\n",
       "      'http://biothings.io/explorer/vocab/terms/reactome.pathway/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/biocarta.pathway/',\n",
       "       'template': '/query?q=pathway.biocarta.id:{pathwayid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/kegg.pathway/',\n",
       "       'template': '/query?q=pathway.kegg.id:{pathwayid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/wikipathways/',\n",
       "       'template': '/query?q=pathway.wikipathways.id:{pathwayid}&size=100'},\n",
       "      {'valueType': 'http://identifiers.org/pharmgkb.pathways/',\n",
       "       'template': '/query?q=pathway.pharmgkb.id:{pathwayid}&size=100'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.pathway/',\n",
       "       'template': '/query?q=pathway.reactome.id:{pathwayid}&size=100'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': 'hits.entrezgene',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'hits.symbol',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'mygene.info/jsonld_context/mygene_query_pathway.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mygene.info/jsonld_context/mygene_query_pathway.json',\n",
       "  'output': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/ncbigene/': {'assoc:PathwayToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PathwayToGeneAssociation'],\n",
       "  'input': ['http://identifiers.org/biocarta.pathway/',\n",
       "   'http://identifiers.org/kegg.pathway/',\n",
       "   'http://identifiers.org/wikipathways/',\n",
       "   'http://identifiers.org/pharmgkb.pathways/',\n",
       "   'http://biothings.io/explorer/vocab/terms/reactome.pathway/'],\n",
       "  'api': 'MyGene.info API',\n",
       "  'server': 'http://mygene.info/v3'},\n",
       " 'http://mygene.info/v3/querygo': {'get': {'summary': 'Make gene query and return matching gene hits',\n",
       "   'parameters': [{'name': 'goid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"CDK2\", \"NM_052827\", \"204639_at\". The detailed query syntax can be found at http://docs.mygene.info/en/latest/doc/query_service.html',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/go/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/go/',\n",
       "       'template': '/query?q=GO:{goid}&size=500'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': 'hits.entrezgene',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'hits.symbol',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'mygene.info/jsonld_context/mygene_query_pathway.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mygene.info/jsonld_context/mygene_query_pathway.json',\n",
       "  'output': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/ncbigene/': {'assoc:PathwayToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PathwayToGeneAssociation'],\n",
       "  'input': ['http://identifiers.org/go/'],\n",
       "  'api': 'MyGene.info API',\n",
       "  'server': 'http://mygene.info/v3'},\n",
       " 'http://mygene.info/v3/gene/{geneid}': {'get': {'summary': 'For a given gene id, return the matching gene object',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'Entrez or Ensembl gene id, e.g., 1017, ENSG00000170248. A retired Entrez Gene id works too if it is replaced by a new one, e.g., 245794',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://biothings.io/explorer/vocab/terms/ensembl.gene/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/gene/{geneid}'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.gene/',\n",
       "       'template': '/gene/{geneid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A gene object',\n",
       "     'x-responseValueType': [{'path': 'symbol',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'},\n",
       "      {'path': 'umls.cui',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.gene/'},\n",
       "      {'path': 'unigene', 'valueType': 'http://identifiers.org/unigene/'},\n",
       "      {'path': 'uniprot.Swiss-Prot',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'},\n",
       "      {'path': 'pdb', 'valueType': 'http://identifiers.org/pdb/'},\n",
       "      {'path': 'pathway.biocarta.id',\n",
       "       'valueType': 'http://identifiers.org/biocarta.pathway/'},\n",
       "      {'path': 'pathway.kegg.id',\n",
       "       'valueType': 'http://identifiers.org/kegg.pathway/'},\n",
       "      {'path': 'pathway.reactome.id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.pathway/'},\n",
       "      {'path': 'pathway.wikipathways.id',\n",
       "       'valueType': 'http://identifiers.org/wikipathways/'},\n",
       "      {'path': 'pathway.pharmgkb.id',\n",
       "       'valueType': 'http://identifiers.org/pharmgkb.pathways/'},\n",
       "      {'path': 'go.bp.id', 'valueType': 'http://identifiers.org/go/'},\n",
       "      {'path': 'ensembl.gene',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.gene/'},\n",
       "      {'path': 'ensembl.protein',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.protein/'},\n",
       "      {'path': 'ensembl.transcript',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.transcript/'},\n",
       "      {'path': 'MIM',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.gene/'},\n",
       "      {'path': 'HGNC', 'valueType': 'http://identifiers.org/hgnc/'},\n",
       "      {'path': 'miRBase', 'valueType': 'http://identifiers.org/mirbase/'}],\n",
       "     'x-JSONLDContext': 'mygene.info/jsonld_context/mygene_gene_object.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mygene.info/jsonld_context/mygene_gene_object.json',\n",
       "  'output': ['http://identifiers.org/hgnc.symbol/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.gene/',\n",
       "   'http://identifiers.org/unigene/',\n",
       "   'http://identifiers.org/uniprot/',\n",
       "   'http://identifiers.org/pdb/',\n",
       "   'http://identifiers.org/biocarta.pathway/',\n",
       "   'http://identifiers.org/kegg.pathway/',\n",
       "   'http://biothings.io/explorer/vocab/terms/reactome.pathway/',\n",
       "   'http://identifiers.org/wikipathways/',\n",
       "   'http://identifiers.org/pharmgkb.pathways/',\n",
       "   'http://identifiers.org/go/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.gene/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.protein/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.transcript/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.gene/',\n",
       "   'http://identifiers.org/hgnc/',\n",
       "   'http://identifiers.org/mirbase/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/biocarta.pathway/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'},\n",
       "               'http://identifiers.org/kegg.pathway/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/reactome.pathway/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'},\n",
       "               'http://identifiers.org/wikipathways/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'},\n",
       "               'http://identifiers.org/pharmgkb.pathways/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'},\n",
       "               'http://identifiers.org/uniprot/': {'assoc:GeneToGeneProductAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.gene/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/ensembl.protein/': {'assoc:GeneToGeneProductAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/ensembl.gene/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/ensembl.transcript/': {'assoc:GeneToGeneProductAssociation'},\n",
       "               'http://identifiers.org/go/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation',\n",
       "                'assoc:MacromolecularMachineToCellularComponentAssociation',\n",
       "                'assoc:MacromolecularMachineToMolecularActivityAssociation'},\n",
       "               'http://identifiers.org/pdb/': {'assoc:GeneToProteinStructureAssociation'},\n",
       "               'http://identifiers.org/unigene/': {'assoc:GeneToGeneProductAssociation'},\n",
       "               'http://identifiers.org/hgnc.symbol/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/omim.gene/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/hgnc/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/mirbase/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/gene-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneToProteinStructureAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/MacromolecularMachineToBiologicalProcessAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/MacromolecularMachineToCellularComponentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/MacromolecularMachineToMolecularActivityAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/GeneToGeneProductAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ensembl.gene/'],\n",
       "  'api': 'MyGene.info API',\n",
       "  'server': 'http://mygene.info/v3'},\n",
       " 'http://MyChem.info/v1/querybydrugtarget': {'get': {'summary': 'Query for chemicals which interacts with proteins',\n",
       "   'parameters': [{'name': 'uniprotid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string.',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/uniprot/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uniprot/',\n",
       "       'template': '/query?q=drugbank.targets.uniprot:{uniprotid} OR drugcentral.bioactivity.uniprot_id:{uniprotid} OR pharmgkb.cross_references.uniprotkb:{uniprotid} & fields=_id'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': '_id',\n",
       "       'valueType': 'http://identifiers.org/inchikey/'}],\n",
       "     'x-JSONLDContext': 'mychem.info/jsonld_context/mychem_querybydrugtarget.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mychem.info/jsonld_context/mychem_querybydrugtarget.json',\n",
       "  'output': ['http://identifiers.org/inchikey/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/inchikey/': {'assoc:GeneOrGeneProductToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/uniprot/'],\n",
       "  'api': 'MyChem.info API',\n",
       "  'server': 'http://MyChem.info/v1'},\n",
       " 'http://MyChem.info/v1/querybydiseaseid': {'get': {'summary': 'Query disease to chemical',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"Fludara\", \"CHEMBI:63599\", \"DB01073\".',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/snomedct/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/snomedct/',\n",
       "       'template': '/query?q=drugcentral.drug_use.snomed_id:{diseaseid}&fields=_id'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': '_id',\n",
       "       'valueType': 'http://identifiers.org/inchikey/'}],\n",
       "     'x-JSONLDContext': 'mychem.info/jsonld_context/mychem_query_disease2chemical.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mychem.info/jsonld_context/mychem_query_disease2chemical.json',\n",
       "  'output': ['http://identifiers.org/inchikey/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/inchikey/': {'assoc:DiseaseToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/snomedct/'],\n",
       "  'api': 'MyChem.info API',\n",
       "  'server': 'http://MyChem.info/v1'},\n",
       " 'http://MyChem.info/v1/querybydrugid': {'get': {'summary': 'Make chemical query and return matching chemical hits',\n",
       "   'parameters': [{'name': 'drugid',\n",
       "     'in': 'path',\n",
       "     'description': 'Query string. Examples \"Fludara\", \"CHEMBI:63599\", \"DB01073\".',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/rxcui/',\n",
       "      'http://identifiers.org/chebi/',\n",
       "      'http://identifiers.org/chembl.compound/',\n",
       "      'http://identifiers.org/drugbank/',\n",
       "      'http://identifiers.org/iuphar.ligand/',\n",
       "      'http://identifiers.org/kegg.drug/',\n",
       "      'http://identifiers.org/kegg.compound/',\n",
       "      'http://identifiers.org/pubchem.compound/',\n",
       "      'http://identifiers.org/unii/',\n",
       "      'http://identifiers.org/inchi/',\n",
       "      'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "      'http://identifiers.org/pharmgkb.drug/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/rxcui/',\n",
       "       'template': 'aeolus.rxcui:input_value OR drugcentral.xref.rxnorm:input_value OR ginas.xrefs.RXCUI:input_value OR pharmgkb.external_vocabulary.rxnorm:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/chebi/',\n",
       "       'template': 'drugcentral.xref.chebi:input_value OR chebi.chebi_id:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/chembl.compound/',\n",
       "       'template': '/query?q=chembl.molecule_chembl_id:{drugid} OR drugbank.chembl:{drugid} OR drugcentral.xref.chembl_id:{drugid}&fields=drugcentral,drugbank,chembl'},\n",
       "      {'valueType': 'http://identifiers.org/drugbank/',\n",
       "       'template': '/query?q=drugcentral.xref.drugbank_id:{drugid} OR pharmgkb.cross_references.drugbank:{drugid} OR drugbank.drugbank_id:{drugid} OR chebi.drugbank_database_links:{drugid}'},\n",
       "      {'valueType': 'http://identifiers.org/iuphar.ligand/',\n",
       "       'template': 'drugcentral.xref.iuphar_ligand_id:input_value OR pharmgkb.cross_references.iuphar_ligand:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/kegg.drug/',\n",
       "       'template': 'drugcentral.xref.kegg_drug:input_value OR chebi.kegg_drug_database_links:input_value OR drugbank.kegg_drug:input_value OR pharmgkb.cross_references.kegg_drug:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/kegg.compound/',\n",
       "       'template': 'chebi.kegg_compound_database_links:input_value OR pharmgkb.cross_references.kegg_compound:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/pubchem.compound/',\n",
       "       'template': 'drugcentral.xref.pubchem_cid:input_value OR chebi.pubchem_database_links:input_value OR pharmgkb.cross_references.pubchem_compound:input_value OR pubchem.cid:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/unii/',\n",
       "       'template': 'drugcentral.xref.unii:input_value OR aeolus.unii:input_value OR ginas.unii:input_value OR unii.unii:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/inchikey/',\n",
       "       'template': 'aeolus.inchikey:input_value OR chebi.inchikey:input_value OR chembl.inchi_key:input_value OR drugbank.inchi_key:input_value OR drugcentral.structures.inchikey:input_value OR ginas.inchikey:input_value OR pubchem.inchi_key:input_value OR unii.inchikey:input_value'},\n",
       "      {'valueType': 'http://identifiers.org/inchi/',\n",
       "       'template': 'pubchem.inchi:input_value OR pharmgkb.inchi:input_value OR drugcentral.structures.inchi:input_value OR drugbank.inchi:input_value OR chembl.inchi:input_value OR chebi.inchi:input_value'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "       'template': 'aeolus.drug_name:input_value OR chebi.chebi_name:input_value OR chembl.pref_name:input_value OR drugbank.name:input_value & fields=drugcentral, chebi, drugbank'},\n",
       "      {'valueType': 'http://identifiers.org/pharmgkb.drug/',\n",
       "       'template': 'pharmgkb.pharmgkb_accession_id:input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"hits\" property',\n",
       "     'x-responseValueType': [{'path': '_id',\n",
       "       'valueType': 'http://identifiers.org/inchikey/'},\n",
       "      {'path': 'aeolus.drug_name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'},\n",
       "      {'path': 'drugbank.accession_number',\n",
       "       'valueType': 'http://identifiers.org/drugbank/'},\n",
       "      {'path': 'chembl',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'},\n",
       "      {'path': 'pubchem',\n",
       "       'valueType': 'http://identifiers.org/pubchem.compound/'},\n",
       "      {'path': 'drugbank.target.uniprot',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'},\n",
       "      {'path': 'drugcentral.xref.umls',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.compound/'},\n",
       "      {'path': 'drugcentral.xref.mesh',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mesh.compound/'},\n",
       "      {'path': 'chebi.definition',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/chemical-description/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.snomed_id',\n",
       "       'valueType': 'http://identifiers.org/snomedct/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.umls_cui',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.concept_name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'},\n",
       "      {'path': 'drugcental.fda_adverse_event.meddra_code',\n",
       "       'valueType': 'http://identifiers.org/meddra/'},\n",
       "      {'path': 'drugcentral.pharmacology_class.fda_epc.code',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/ndfrt/'},\n",
       "      {'path': 'chembl.max_phase',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-phase/'}],\n",
       "     'x-JSONLDContext': 'mychem.info/jsonld_context/mychem_query_object.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mychem.info/jsonld_context/mychem_query_object.json',\n",
       "  'output': ['http://identifiers.org/inchikey/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/drugbank/',\n",
       "   'http://identifiers.org/chembl.compound/',\n",
       "   'http://identifiers.org/pubchem.compound/',\n",
       "   'http://identifiers.org/uniprot/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.compound/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mesh.compound/',\n",
       "   'http://biothings.io/explorer/vocab/terms/chemical-description/',\n",
       "   'http://identifiers.org/snomedct/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "   'http://identifiers.org/meddra/',\n",
       "   'http://biothings.io/explorer/vocab/terms/ndfrt/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-phase/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/inchikey/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/rxcui/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/chembl.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/chebi/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/drugbank/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/iuphar.ligand/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/kegg.drug/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/pubchem.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/unii/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/mesh.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/inchi/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.disease/': {'assoc:Treats'},\n",
       "               'http://identifiers.org/meddra/': {'assoc:Causes'},\n",
       "               'http://biothings.io/explorer/vocab/terms/ndfrt/': {'assoc:Subclass_of'},\n",
       "               'http://identifiers.org/pharmgkb.drug/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/kegg.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/pubchem.substance/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/clinicaltrials/': {'assoc:ChemicalToClinicalTrialAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/drug-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/chemical-description/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/uniprot/': {'assoc:Regulates'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/Regulates',\n",
       "   'http://biothings.io/explorer/vocab/objects/Subclass_of',\n",
       "   'http://biothings.io/explorer/vocab/objects/ChemicalToClinicalTrialAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/Treats',\n",
       "   'http://biothings.io/explorer/vocab/objects/Causes'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/rxcui/',\n",
       "   'http://identifiers.org/chebi/',\n",
       "   'http://identifiers.org/chembl.compound/',\n",
       "   'http://identifiers.org/drugbank/',\n",
       "   'http://identifiers.org/iuphar.ligand/',\n",
       "   'http://identifiers.org/kegg.drug/',\n",
       "   'http://identifiers.org/kegg.compound/',\n",
       "   'http://identifiers.org/pubchem.compound/',\n",
       "   'http://identifiers.org/unii/',\n",
       "   'http://identifiers.org/inchi/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/pharmgkb.drug/'],\n",
       "  'api': 'MyChem.info API',\n",
       "  'server': 'http://MyChem.info/v1'},\n",
       " 'http://MyChem.info/v1/drug/{drugid}': {'get': {'tags': ['chemical'],\n",
       "   'summary': 'Retrieve chemical objects based on ID',\n",
       "   'parameters': [{'name': 'drugid',\n",
       "     'in': 'path',\n",
       "     'description': 'DrugBank ID, InCHIKey, CHEBI ID, CHEMBL ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/inchikey/',\n",
       "      'http://identifiers.org/drugbank/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/inchikey/',\n",
       "       'template': '/drug/{drugid}'},\n",
       "      {'valueType': 'http://identifiers.org/drugbank/',\n",
       "       'template': '/drug/{drugid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching chemical object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': '_id',\n",
       "       'valueType': 'http://identifiers.org/inchikey/'},\n",
       "      {'path': 'aeolus.drugname',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'},\n",
       "      {'path': 'pubchem.inchi', 'valueType': 'http://identifiers.org/inchi/'},\n",
       "      {'path': 'drugcentral.xref.unii',\n",
       "       'valueType': 'http://identifiers.org/unii/'},\n",
       "      {'path': 'drugbank.enzymes.uniprot',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'},\n",
       "      {'path': 'drugbank.drugbank_id',\n",
       "       'valueType': 'http://identifiers.org/drugbank/'},\n",
       "      {'path': 'cadd.gene.gene_id',\n",
       "       'valueType': 'http://identifiers.org/dbsnp/'},\n",
       "      {'path': 'cadd.gene.feature_id',\n",
       "       'valueType': 'http://identifiers.org/pubmed/'},\n",
       "      {'path': 'cadd.gene.ccds_id',\n",
       "       'valueType': 'http://identifiers.org/clinicaltrials/'},\n",
       "      {'path': 'aeolus.rxcui',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/rxcui/'},\n",
       "      {'path': 'chebi.chebi_id', 'valueType': 'http://identifiers.org/chebi/'},\n",
       "      {'path': 'chembl.molecule_chembl_id',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'},\n",
       "      {'path': 'drugcentral.xref.kegg_drug',\n",
       "       'valueType': 'http://identifiers.org/kegg.drug/'},\n",
       "      {'path': 'chebi.kegg_compound_database_links',\n",
       "       'valueType': 'http://identifiers.org/kegg.compound/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.snomed_id',\n",
       "       'valueType': 'http://identifiers.org/snomedct/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.umls_cui',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/'},\n",
       "      {'path': 'drugcentral.xref.iuphar_ligand_id',\n",
       "       'valueType': 'http://identifiers.org/iuphar.ligand/'},\n",
       "      {'path': 'drugcentral.drug_use.indication.concept_name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'},\n",
       "      {'path': 'pharmgkb.pharmgkb_accession_id',\n",
       "       'valueType': 'http://identifiers.org/pharmgkb.drug/'},\n",
       "      {'path': 'drugcentral.xref.pubchem_cid',\n",
       "       'valueType': 'http://identifiers.org/pubchem.compound/'},\n",
       "      {'path': 'pharmgkb.cross_references.pubchem_substance',\n",
       "       'valueType': 'http://identifiers.org/pubchem.substance/'},\n",
       "      {'path': 'drugcentral.xref.umls',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.compound/'},\n",
       "      {'path': 'drugcentral.xref.mesh',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mesh.compound/'},\n",
       "      {'path': 'chebi.definition',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/chemical-description/'}],\n",
       "     'x-JSONLDContext': 'mychem.info/jsonld_context/mychem_drug_object.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mychem.info/jsonld_context/mychem_drug_object.json',\n",
       "  'output': ['http://identifiers.org/inchikey/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/inchi/',\n",
       "   'http://identifiers.org/unii/',\n",
       "   'http://identifiers.org/uniprot/',\n",
       "   'http://identifiers.org/drugbank/',\n",
       "   'http://identifiers.org/dbsnp/',\n",
       "   'http://identifiers.org/pubmed/',\n",
       "   'http://identifiers.org/clinicaltrials/',\n",
       "   'http://biothings.io/explorer/vocab/terms/rxcui/',\n",
       "   'http://identifiers.org/chebi/',\n",
       "   'http://identifiers.org/chembl.compound/',\n",
       "   'http://identifiers.org/kegg.drug/',\n",
       "   'http://identifiers.org/kegg.compound/',\n",
       "   'http://identifiers.org/snomedct/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://identifiers.org/iuphar.ligand/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "   'http://identifiers.org/pharmgkb.drug/',\n",
       "   'http://identifiers.org/pubchem.compound/',\n",
       "   'http://identifiers.org/pubchem.substance/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.compound/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mesh.compound/',\n",
       "   'http://biothings.io/explorer/vocab/terms/chemical-description/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/inchikey/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/drug-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/rxcui/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/chebi/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/chembl.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/drugbank/': {'assoc:EquivalentAssociation',\n",
       "                'assoc:PairwiseChemicalInteractionAssociation'},\n",
       "               'http://identifiers.org/iuphar.ligand/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/kegg.drug/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/pubchem.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/unii/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/mesh.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/inchi/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.disease/': {'assoc:treats'},\n",
       "               'http://identifiers.org/pharmgkb.drug/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/kegg.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/pubchem.substance/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/chemical-description/': {'assoc:HasDescriptionAssociation'},\n",
       "               'http://identifiers.org/uniprot/': {'assoc:ChemicalToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PairwiseChemicalInteractionAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/treats',\n",
       "   'http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/ChemicalToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://identifiers.org/inchikey/',\n",
       "   'http://identifiers.org/drugbank/'],\n",
       "  'api': 'MyChem.info API',\n",
       "  'server': 'http://MyChem.info/v1'},\n",
       " 'http://mydisease.info/v1/disease/{mondoid}': {'get': {'summary': 'Retrieve disease objects based on mondo ID',\n",
       "   'parameters': [{'name': 'mondoid',\n",
       "     'in': 'path',\n",
       "     'description': 'MONDO ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/disease/MONDO:{mondoid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching disease object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'mondo.label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'},\n",
       "      {'path': 'mondo.definition',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-description/'},\n",
       "      {'path': 'mondo.xrefs.doid',\n",
       "       'valueType': 'http://identifiers.org/doid/'},\n",
       "      {'path': 'mondo.xrefs.omim',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/'},\n",
       "      {'path': 'mondo.xrefs.umls',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_disease_annotation.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_disease_annotation.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-description/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.disease/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/doid/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/omim.disease/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/umls.disease/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/disease-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/disease-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://mydisease.info/v1/querydiseasetogene': {'get': {'summary': 'Retrieve genes related to disease',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'disease ID, e.g. UMLS ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "      'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "      'http://identifiers.org/doid/',\n",
       "      'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "       'template': '/query?q=disgenet.xrefs.umls:{diseaseid}&fields=disgenet.genes_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/query?q=_id:\"MONDO:{diseaseid}\"&fields=disgenet.genes_related_to_disease'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/query?q=mondo.xrefs.doid:\"DOID:{diseaseid}\"&fields=disgenet.genes_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/query?q=mondo.xrefs.omim:{diseaseid}&fields=disgenet.genes_related_to_disease'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching gene object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'hits.disgenet.genes_related_to_disease.gene_id',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'hits.disgenet.genes_related_to_disease.gene_name',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_disease2gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_disease2gene.json',\n",
       "  'output': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/ncbigene/': {'assoc:DiseaseToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://mydisease.info/v1/querydiseasetovariant': {'get': {'summary': 'Retrieve variants related to disease',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'disease ID, e.g. UMLS ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "      'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "      'http://identifiers.org/doid/',\n",
       "      'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "       'template': '/query?q=disgenet.xrefs.umls:{diseaseid}&fields=disgenet.variants_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/query?q=_id:\"MONDO:{diseaseid}\"&fields=disgenet.variants_related_to_disease'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/query?q=mondo.xrefs.doid:\"DOID:{diseaseid}\"&fields=disgenet.variants_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/query?q=mondo.xrefs.omim:{diseaseid}&fields=disgenet.variants_related_to_disease'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching gene object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'hits.disgenet.variants_related_to_disease.rsid',\n",
       "       'valueType': 'http://identifiers.org/dbsnp/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_disease2variant.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_disease2variant.json',\n",
       "  'output': ['http://identifiers.org/dbsnp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/dbsnp/': {'assoc:DiseaseToVariantAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToVariantAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://mydisease.info/v1/querydiseasetophenotype': {'get': {'summary': 'Retrieve phenotypes related to disease',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'disease ID, e.g. UMLS ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "      'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "      'http://identifiers.org/doid/',\n",
       "      'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "      'http://identifiers.org/orphanet/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "       'template': '/query?q=mondo.xrefs.umls:{diseaseid}&fields=hpo.phenotype_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/query?q=_id:\"MONDO:{diseaseid}\"&fields=hpo.phenotype_related_to_disease'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/query?q=mondo.xrefs.doid:\"DOID:{diseaseid}\"&fields=hpo.phenotype_related_to_disease'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/query?q=mondo.xrefs.omim:{diseaseid}&fields=hpo.phenotype_related_to_disease'},\n",
       "      {'valueType': 'http://identifiers.org/orphanet/',\n",
       "       'template': '/query?q=mondo.xrefs.orphanet:{diseaseid}&fields=hpo.phenotype_related_to_disease'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching gene object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'hits.hpo.phenotypes_related_to_disease.rsid',\n",
       "       'valueType': 'http://identifiers.org/hp/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_disease2phenotype.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_disease2phenotype.json',\n",
       "  'output': ['http://identifiers.org/hp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hp/': {'assoc:DiseaseToPhenotypicFeatureAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://identifiers.org/orphanet/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://mydisease.info/v1/queryphenotypetodisease': {'get': {'summary': 'Retrieve diseases related to phenotype',\n",
       "   'parameters': [{'name': 'phenotypeid',\n",
       "     'in': 'path',\n",
       "     'description': 'phenotype ID, e.g. HPO ID',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/hp/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hp/',\n",
       "       'template': '/query?q=hpo.phenotype_related_to_disease.hpo_id:\"HP:{phenotypeid}\"&fields=_id,mondo'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching disease object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'hits.mondo.xrefs.doid',\n",
       "       'valueType': 'http://identifiers.org/doid/'},\n",
       "      {'path': 'hits.mondo.xrefs.omim',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_phenotype2disease.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_phenotype2disease.json',\n",
       "  'output': ['http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/doid/': {'assoc:PhenotypicFeatureToDiseaseAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/omim.disease/': {'assoc:PhenotypicFeatureToDiseaseAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PhenotypicFeatureToDiseaseAssociation'],\n",
       "  'input': ['http://identifiers.org/hp/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://mydisease.info/v1/querygenetodisease': {'get': {'summary': 'Retrieve diseases related to phenotype',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'gene ID, e.g. ncbigene id',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://identifiers.org/hgnc.symbol/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/query?q=disgenet.genes_related_to_disease.gene_id:{geneid}&fields=_id,mondo'},\n",
       "      {'valueType': 'http://identifiers.org/hgnc.symbol/',\n",
       "       'template': '/query?q=disgenet.genes_related_to_disease.gene_name:{geneid}&fields=_id,mondo'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A matching disease object',\n",
       "     'content': {'application/json': {'schema': {'$ref': '#/components/schemas/Chem'}}},\n",
       "     'x-responseValueType': [{'path': 'hits.mondo.xrefs.doid',\n",
       "       'valueType': 'http://identifiers.org/doid/'},\n",
       "      {'path': 'hits.mondo.xrefs.omim',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/'}],\n",
       "     'x-JSONLDContext': 'mydisease.info/jsonld_context/mydisease_gene2disease.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/mydisease.info/jsonld_context/mydisease_gene2disease.json',\n",
       "  'output': ['http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/doid/': {'assoc:GeneOrGeneProductToDiseaseAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/omim.disease/': {'assoc:GeneOrGeneProductToDiseaseAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToDiseaseAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc.symbol/'],\n",
       "  'api': 'MyDisease.info API',\n",
       "  'server': 'http://mydisease.info/v1'},\n",
       " 'http://dgidb.genome.wustl.edu/api/v2/interactions.json?drugs={drugname}': {'get': {'summary': 'return interactions for a drug with a gene',\n",
       "   'parameters': [{'name': 'drugname',\n",
       "     'in': 'path',\n",
       "     'description': 'Put gene names here, e.g. riluzole ',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "      'http://identifiers.org/chembl.compound/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "       'template': '/interactions.json?drugs={drugname}'},\n",
       "      {'valueType': 'http://identifiers.org/chembl.compound/',\n",
       "       'template': '/interactions.json?drugs={drugname}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"matchedTerm\" property',\n",
       "     'x-responseValueType': [{'path': 'mathcedTerms.interactions',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'}],\n",
       "     'x-JSONLDContext': 'dgidb/jsonld_context/dgidb_drug_gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/dgidb/jsonld_context/dgidb_drug_gene.json',\n",
       "  'output': ['http://identifiers.org/hgnc.symbol/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc.symbol/': {'assoc:ChemicalToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ChemicalToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/chembl.compound/'],\n",
       "  'api': 'DGIdb API',\n",
       "  'server': 'http://dgidb.genome.wustl.edu/api/v2'},\n",
       " 'http://dgidb.genome.wustl.edu/api/v2/interactions.json?genes={genesymbol}': {'get': {'summary': 'return interactions for a gene with a drug',\n",
       "   'parameters': [{'name': 'genesymbol',\n",
       "     'in': 'path',\n",
       "     'description': 'Put gene names here, e.g. FLT1, CXCR4 ',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/hgnc.symbol/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hgnc.symbol/',\n",
       "       'template': '/interactions.json?genes={genesymbol}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"matchedTerm\" property',\n",
       "     'x-responseValueType': [{'path': 'matchedTerms.interactions.drugName',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'},\n",
       "      {'path': 'matchedTerms.interactions.chemblId',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'}],\n",
       "     'x-JSONLDContext': 'dgidb/jsonld_context/dgidb_gene_drug.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/dgidb/jsonld_context/dgidb_gene_drug.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/chembl.compound/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/chembl.compound/': {'assoc:GeneOrGeneProductToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/hgnc.symbol/'],\n",
       "  'api': 'DGIdb API',\n",
       "  'server': 'http://dgidb.genome.wustl.edu/api/v2'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/phenotype/{phenotypeid}': {'get': {'summary': 'Returns phenotype label',\n",
       "   'parameters': [{'name': 'phenotypeid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of phenotype, e.g. HP:000001. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/hp/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hp/',\n",
       "       'template': '/bioentity/phenotype/HP:{phenotypeid}'}]}],\n",
       "   'responses': {'200': {'description': 'phenotype label',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/phenotype-name/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_phenotype_label.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_phenotype_label.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/phenotype-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/phenotype-name/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://identifiers.org/hp/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/disease/{diseaseid}': {'get': {'summary': 'Returns phenotype label',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of diseae, e.g. MONDO:00001. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "      'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "      'http://identifiers.org/doid/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/bioentity/disease/MONDO:{diseaseid}'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/bioentity/disease/OMIM:{diseaseid}'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/bioentity/disease/DOID:{diseaseid}'}]}],\n",
       "   'responses': {'200': {'description': 'disease label',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_disease_label.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_disease_label.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/disease-name/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://identifiers.org/doid/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/expression/anatomy': {'get': {'summary': 'Returns anatomy associated with a gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:4750. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/expression/anatomy?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/uberon/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_anatomy.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_anatomy.json',\n",
       "  'output': ['http://identifiers.org/uberon/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uberon/': {'assoc:GeneOrGeneProductToAnatomicalEntityAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToAnatomicalEntityAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/disease/{diseaseid}/genes': {'get': {'summary': 'Returns genes associated with a disease',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of disease, e.g. OMIM:605543, DOID:678. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "      'http://identifiers.org/doid/',\n",
       "      'http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/bioentity/disease/OMIM:{diseaseid}/genes?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/bioentity/disease/DOID:{diseaseid}/genes?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/bioentity/disease/MONDO:{diseaseid}/genes?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hgnc/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_disease_gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_disease_gene.json',\n",
       "  'output': ['http://identifiers.org/hgnc/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc/': {'assoc:DiseaseToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/disease/{diseaseid}/phenotypes': {'get': {'summary': 'Returns phenotypes associated with a disease',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of disease, e.g. OMIM:605543, DOID:678. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "      'http://identifiers.org/doid/',\n",
       "      'http://identifiers.org/orphanet/',\n",
       "      'http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "       'template': '/bioentity/disease/OMIM:{diseaseid}/phenotypes/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/orphanet/',\n",
       "       'template': '/bioentity/disease/Orphanet:{diseaseid}/phenotypes/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/bioentity/disease/DOID:{diseaseid}/phenotypes/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/',\n",
       "       'template': '/bioentity/disease/MONDO:{diseaseid}/phenotypes/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hp/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_disease_phenotype.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_disease_phenotype.json',\n",
       "  'output': ['http://identifiers.org/hp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hp/': {'assoc:DiseaseToPhenotypicFeatureAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseToPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://identifiers.org/orphanet/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/homologs': {'get': {'summary': 'Returns homologs for a gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:4750. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/homologs/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'edge.obj',\n",
       "       'valueType': 'http://identifiers.org/mgi/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/zfin/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/hgnc/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/fb/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/xenbase/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/rgd/'},\n",
       "      {'path': 'edge.obj', 'valueType': 'http://identifiers.org/sgd/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_homolog.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_homolog.json',\n",
       "  'output': ['http://identifiers.org/mgi/',\n",
       "   'http://identifiers.org/zfin/',\n",
       "   'http://identifiers.org/hgnc/',\n",
       "   'http://identifiers.org/fb/',\n",
       "   'http://identifiers.org/xenbase/',\n",
       "   'http://identifiers.org/rgd/',\n",
       "   'http://identifiers.org/sgd/'],\n",
       "  'relation': {'http://identifiers.org/mgi/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/zfin/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/hgnc/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/fb/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/xenbase/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/rgd/': {'assoc:GeneToGeneHomologyAssociation'},\n",
       "   'http://identifiers.org/sgd/': {'assoc:GeneToGeneHomologyAssociation'}},\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneToGeneHomologyAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/interactions': {'get': {'summary': 'Returns interactions for a gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:4750. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://identifiers.org/hgnc/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/interactions/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/hgnc/',\n",
       "       'template': '/bioentity/gene/HGNC:{geneid}/interactions/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hgnc/'},\n",
       "      {'path': 'objects', 'valueType': 'http://identifiers.org/mgi/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_interaction.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_interaction.json',\n",
       "  'output': ['http://identifiers.org/hgnc/', 'http://identifiers.org/mgi/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc/': {'assoc:PairwiseGeneOrProteinInteractionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PairwiseGeneOrProteinInteractionAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/phenotypes': {'get': {'summary': 'Returns phenotypes associated with gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:4750. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://identifiers.org/hgnc/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/phenotypes/?exclude_automatic_assertions=true&unselect_evidence=true'},\n",
       "      {'valueType': 'http://identifiers.org/hgnc/',\n",
       "       'template': '/bioentity/gene/HGNC:{geneid}/phenotypes/?exclude_automatic_assertions=true&unselect_evidence=true'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hp/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_phenotype.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_phenotype.json',\n",
       "  'output': ['http://identifiers.org/hp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hp/': {'assoc:GeneOrGeneProductToPhenotypicFeatureAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/pathways': {'get': {'summary': 'Returns pathways associated with gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:4750. Equivalent IDs can be used with same results',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://identifiers.org/hgnc/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/pathways/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/hgnc/',\n",
       "       'template': '/bioentity/gene/HGNC:{geneid}/pathways/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.pathway/'},\n",
       "      {'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/kegg.pathway/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_pathway.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_pathway.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/reactome.pathway/',\n",
       "   'http://identifiers.org/kegg.pathway/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/reactome.pathway/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/MacromolecularMachineToBiologicalProcessAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/pathway/{pathwayid}/genes': {'get': {'summary': 'Returns genes associated with a pathway',\n",
       "   'parameters': [{'name': 'pathwayid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE any pathway element. E.g. REACT:R-HSA-5387390',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/reactome/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/reactome/',\n",
       "       'template': '/bioentity/pathway/REACT:{pathwayid}/genes/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hgnc/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_pathway_gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_pathway_gene.json',\n",
       "  'output': ['http://identifiers.org/hgnc/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc/': {'assoc:PathwayToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PathwayToGeneAssociation'],\n",
       "  'input': [],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/variant/{variantid}/genes': {'get': {'summary': 'Returns genes associated with a variant',\n",
       "   'parameters': [{'name': 'variantid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of variant, e.g. ZFIN:ZDB-ALT-010427-8, ClinVarVariant:39783',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/clinvar/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/clinvar/',\n",
       "       'template': '/bioentity/variant/ClinVarVariant:{variantid}/genes/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hgnc/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_variant_gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_variant_gene.json',\n",
       "  'output': ['http://identifiers.org/hgnc/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc/': {'assoc:VariantToGeneAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/VariantToGeneAssociation'],\n",
       "  'input': [],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/anatomy/{anatomyid}/genes': {'get': {'summary': 'Returns genes associated with an anatomy',\n",
       "   'parameters': [{'name': 'anatomyid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of anatomy, e.g. UBERON:0002037',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/uberon/',\n",
       "      'http://identifiers.org/cl/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uberon/',\n",
       "       'template': '/bioentity/anatomy/UBERON:{anatomyid}/genes/?fetch_objects=true&rows=250'},\n",
       "      {'valueType': 'http://identifiers.org/cl/',\n",
       "       'template': '/bioentity/anatomy/CL:{anatomyid}/genes/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/hgnc/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_anatomy_gene.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_anatomy_gene.json',\n",
       "  'output': ['http://identifiers.org/hgnc/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc/': {'assoc:AnatomicalEntityToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/AnatomicalEntityToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://identifiers.org/uberon/', 'http://identifiers.org/cl/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/phenotype/{phenotypeid}/anatomy': {'get': {'summary': 'Returns genes associated with an anatomy',\n",
       "   'parameters': [{'name': 'phenotypeid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of phenotype, e.g. HP:0000003',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/hp/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hp/',\n",
       "       'template': '/bioentity/phenotype/HP:{phenotypeid}/anatomy/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://identifiers.org/uberon/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_phenotype_anatomy.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_phenotype_anatomy.json',\n",
       "  'output': ['http://identifiers.org/uberon/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uberon/': {'assoc:DiseaseOrPhenotypicFeatureAssociationToLocationAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseOrPhenotypicFeatureAssociationToLocationAssociation'],\n",
       "  'input': ['http://identifiers.org/hp/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://api.monarchinitiative.org/api/bioentity/gene/{geneid}/diseases': {'get': {'summary': 'Returns diseases associated with a gene',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'CURIE identifier of gene, e.g. NCBIGene:1017',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/',\n",
       "      'http://identifiers.org/hgnc/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/bioentity/gene/NCBIGene:{geneid}/diseases/?fetch_objects=true&rows=200'},\n",
       "      {'valueType': 'http://identifiers.org/hgnc/',\n",
       "       'template': '/bioentity/gene/HGNC:{geneid}/diseases/?fetch_objects=true&rows=250'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'objects',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mondo/'}],\n",
       "     'x-JSONLDContext': 'biolink/jsonld_context/biolink_gene_disease.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/biolink/jsonld_context/biolink_gene_disease.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/mondo/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/mondo/': {'assoc:GeneOrGeneProductToDiseaseAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToDiseaseAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/',\n",
       "   'http://identifiers.org/hgnc/'],\n",
       "  'api': 'BioLink API',\n",
       "  'server': 'https://api.monarchinitiative.org/api'},\n",
       " 'https://www.ebi.ac.uk/chembl/api/data/compound_record': {'get': {'summary': 'return chembl compound id corresponding to a drug name',\n",
       "   'parameters': [{'name': 'compound_name__iexact',\n",
       "     'in': 'query',\n",
       "     'description': 'Put drugname here',\n",
       "     'required': True,\n",
       "     'example': 'riluzole',\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"compound_records\" property',\n",
       "     'x-responseValueType': [{'path': 'compound_records.molecule_chembl_id',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'}],\n",
       "     'x-JSONLDContext': 'ChEMBL/jsonld_context/chembl_drugname2chemblcompound_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ChEMBL/jsonld_context/chembl_drugname2chemblcompound_context.json',\n",
       "  'output': ['http://identifiers.org/chembl.compound/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/chembl.compound/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'api': 'ChEMBL API',\n",
       "  'server': 'https://www.ebi.ac.uk/chembl/api/data'},\n",
       " 'https://www.ebi.ac.uk/chembl/api/data/target_prediction/{chemblid}': {'get': {'summary': 'return predicted binding of a molecule to given biological target',\n",
       "   'parameters': [{'name': 'chemblid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put chembl compound id here',\n",
       "     'required': True,\n",
       "     'example': 'CHEMBL744',\n",
       "     'x-valueType': ['http://identifiers.org/chembl.compound/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/chembl.compound/',\n",
       "       'template': '/target_prediction?molecule_chembl_id__exact={chemblid}&target_organism__exact=Homo sapiens'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"target_prediction\" property',\n",
       "     'x-responseValueType': [{'path': 'target_prediction.target_chembl_id',\n",
       "       'valueType': 'http://identifiers.org/chembl.target/'},\n",
       "      {'path': 'target_prediction.target_accession',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'}],\n",
       "     'x-JSONLDContext': 'ChEMBL/jsonld_context/chembl_target_prediction_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ChEMBL/jsonld_context/chembl_target_prediction_context.json',\n",
       "  'output': ['http://identifiers.org/uniprot/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uniprot/': {'assoc:ChemicalToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ChemicalToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://identifiers.org/chembl.compound/'],\n",
       "  'api': 'ChEMBL API',\n",
       "  'server': 'https://www.ebi.ac.uk/chembl/api/data'},\n",
       " 'https://www.ebi.ac.uk/chembl/api/data/mechanism': {'get': {'summary': 'return drugs interacts with protein',\n",
       "   'parameters': [{'name': 'target_chembl_id',\n",
       "     'in': 'query',\n",
       "     'description': 'Put target ChEMBL ID here, e.g. CHEMBL2326',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/chembl.target/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/chembl.target/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"mechanisms\" property',\n",
       "     'x-responseValueType': [{'path': 'mechanisms.molecule_chembl_id',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'}],\n",
       "     'x-JSONLDContext': 'ChEMBL/jsonld_context/chembl_target2compound_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ChEMBL/jsonld_context/chembl_target2compound_context.json',\n",
       "  'output': ['http://identifiers.org/chembl.compound/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/chembl.compound/': {'assoc:ChemicalToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ChemicalToGeneOrGeneProductAssociation'],\n",
       "  'input': [],\n",
       "  'api': 'ChEMBL API',\n",
       "  'server': 'https://www.ebi.ac.uk/chembl/api/data'},\n",
       " 'https://www.ebi.ac.uk/chembl/api/data/drug_indication': {'get': {'summary': 'return indications for drugs',\n",
       "   'parameters': [{'name': 'molecule_chembl_id',\n",
       "     'in': 'query',\n",
       "     'description': 'Put ChEMBL compound ID here, e.g. CHEMBL112',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/chembl.compound/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/chembl.compound/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"drug_indications\" property',\n",
       "     'x-responseValueType': [{'path': 'drug_indications.mesh_id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mesh.disease/'},\n",
       "      {'path': 'drug_indications.efo_id',\n",
       "       'valueType': 'http://identifiers.org/hp/'},\n",
       "      {'path': 'drug_indications.efo_id',\n",
       "       'valueType': 'http://identifiers.org/efo/'},\n",
       "      {'path': 'drug_indications.efo_id',\n",
       "       'valueType': 'http://identifiers.org/orphanet/'}],\n",
       "     'x-JSONLDContext': 'ChEMBL/jsonld_context/chembl_drugindication_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ChEMBL/jsonld_context/chembl_drugindication_context.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/mesh.disease/',\n",
       "   'http://identifiers.org/hp/',\n",
       "   'http://identifiers.org/efo/',\n",
       "   'http://identifiers.org/orphanet/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/mesh.disease/': {'assoc:ChemicalToDiseaseOrPhenotypicFeatureAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ChemicalToDiseaseOrPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://identifiers.org/chembl.compound/'],\n",
       "  'api': 'ChEMBL API',\n",
       "  'server': 'https://www.ebi.ac.uk/chembl/api/data'},\n",
       " 'http://www.disease-ontology.org/api/metadata/{doid}': {'get': {'summary': 'Returns xref related to a doid',\n",
       "   'parameters': [{'name': 'doid',\n",
       "     'in': 'path',\n",
       "     'description': 'DOID, e.g. DOID:678.',\n",
       "     'required': True,\n",
       "     'schema': {'type': 'string'},\n",
       "     'x-valueType': ['http://identifiers.org/doid/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/metadata/DOID:{doid}'}]}],\n",
       "   'responses': {'200': {'description': 'A distinct set of association.objects (typically ontology terms). This appears at the top level of the results payload',\n",
       "     'x-responseValueType': [{'path': 'xref',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mesh.disease/'},\n",
       "      {'path': 'xref', 'valueType': 'http://identifiers.org/ncit/'},\n",
       "      {'path': 'xref', 'valueType': 'http://identifiers.org/snomedct/'},\n",
       "      {'path': 'xref', 'valueType': 'http://identifiers.org/umls/'},\n",
       "      {'path': 'xref',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/'},\n",
       "      {'path': 'parents', 'valueType': 'http://identifiers.org/doid/'},\n",
       "      {'path': 'name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'}],\n",
       "     'x-JSONLDContext': 'disease ontology api/jsonld_context/do_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/disease ontology api/jsonld_context/do_context.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/mesh.disease/',\n",
       "   'http://identifiers.org/ncit/',\n",
       "   'http://identifiers.org/snomedct/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "  'relation': {'http://biothings.io/explorer/vocab/terms/mesh.disease/': {'assoc:EquivalentAssociation'},\n",
       "   'http://identifiers.org/ncit/': {'assoc:EquivalentAssociation'},\n",
       "   'http://identifiers.org/snomedct/': {'assoc:EquivalentAssociation'},\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/': {'assoc:EquivalentAssociation'},\n",
       "   'http://identifiers.org/doid/': {'assoc:EquivalentAssociation'},\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/': {'assoc:EquivalentAssociation'}},\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://identifiers.org/doid/'],\n",
       "  'api': 'Disease Ontology API',\n",
       "  'server': 'http://www.disease-ontology.org/api'},\n",
       " 'https://reactome.org/ContentService/data/complexes/Uniprot/{uniprotId}': {'get': {'summary': 'Retrieves the list of complexes that contain a given uniprot id.',\n",
       "   'parameters': [{'name': 'uniprotId',\n",
       "     'in': 'path',\n",
       "     'description': 'Put uniprot id here, e.g. P00533',\n",
       "     'required': True,\n",
       "     'example': 'P00533',\n",
       "     'x-valueType': ['http://identifiers.org/uniprot/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uniprot/',\n",
       "       'template': '/data/complexes/UniProt/{uniprotId}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'List of reactome complexes',\n",
       "     'x-responseValueType': [{'path': 'data.stId',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.complex/'}],\n",
       "     'x-JSONLDContext': 'reactome/jsonld_context/uniprot2complex.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/reactome/jsonld_context/uniprot2complex.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/reactome.complex/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/reactome.complex/': {'assoc:isComponentOf'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/isComponentOf'],\n",
       "  'input': ['http://identifiers.org/uniprot/'],\n",
       "  'api': 'Reactome API',\n",
       "  'server': 'https://reactome.org/ContentService'},\n",
       " 'https://reactome.org/ContentService/data/pathways/low/diagram/entity/{ReactomeComplexId}/allForms': {'get': {'summary': 'This method traverses the event hierarchy and retrieves the list of all lower level pathways that have a diagram and contain the given PhysicalEntity in any of its variant forms. These variant forms include for example different post-translationally modified versions of a single protein, or the same chemical in different compartments.',\n",
       "   'parameters': [{'name': 'ReactomeComplexId',\n",
       "     'in': 'path',\n",
       "     'description': 'Put Reactome Complex id here, e.g. R-HSA-199420',\n",
       "     'required': True,\n",
       "     'example': 'R-HSA-199420',\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/reactome.complex/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.complex/',\n",
       "       'template': '/data/pathways/low/diagram/entity/{ReactomeComplexId}/allForms'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'List of reactome pathways',\n",
       "     'x-responseValueType': [{'path': 'data.stId',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.pathway/'}],\n",
       "     'x-JSONLDContext': 'reactome/jsonld_context/complex2pathways.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/reactome/jsonld_context/complex2pathways.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/reactome.pathway/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/reactome.pathway/': {'assoc:MacromolecularMachineToBiologicalProcessAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/MacromolecularMachineToBiologicalProcessAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/reactome.complex/'],\n",
       "  'api': 'Reactome API',\n",
       "  'server': 'https://reactome.org/ContentService'},\n",
       " 'https://reactome.org/ContentService/data/participants/{ReactomePathwaysId}': {'get': {'summary': 'Participants contains a PhysicalEntity (dbId, displayName) and a collection of ReferenceEntities (dbId, name, identifier, url)',\n",
       "   'parameters': [{'name': 'ReactomePathwaysId',\n",
       "     'in': 'path',\n",
       "     'description': 'Put Reactome Pathways id here, e.g. R-HSA-202403',\n",
       "     'required': True,\n",
       "     'example': 'R-HSA-202403',\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/reactome.pathway/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/reactome.pathway/',\n",
       "       'template': '/data/participants/{ReactomePathwaysId}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'List of uniprot Ids which participates in the pathways',\n",
       "     'x-responseValueType': [{'path': 'data.refEntities.identifier',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'}],\n",
       "     'x-JSONLDContext': 'reactome/jsonld_context/pathways2uniprot.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/reactome/jsonld_context/pathways2uniprot.json',\n",
       "  'output': ['http://identifiers.org/uniprot/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uniprot/': {'assoc:BiologicalProcessToMacromolecularMachineAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/BiologicalProcessToMacromolecularMachineAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/reactome.pathway/'],\n",
       "  'api': 'Reactome API',\n",
       "  'server': 'https://reactome.org/ContentService'},\n",
       " 'https://reactome.org/ContentService/interactors/static/molecule/{uniprot}/details': {'get': {'summary': 'Retrieve a detailed interaction information of a given accession',\n",
       "   'parameters': [{'name': 'uniprot',\n",
       "     'in': 'path',\n",
       "     'description': 'Put UniProt id here, e.g. Q06609',\n",
       "     'required': True,\n",
       "     'example': 'Q06609',\n",
       "     'x-valueType': ['http://identifiers.org/uniprot/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uniprot/',\n",
       "       'template': '/interactors/static/molecule/{uniprot}/details'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'List of uniprot Ids which interacts with your input uniprot',\n",
       "     'x-responseValueType': [{'path': 'entities.interactors',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'}],\n",
       "     'x-JSONLDContext': 'reactome/jsonld_context/uniprot2uniprot.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/reactome/jsonld_context/uniprot2uniprot.json',\n",
       "  'output': ['http://identifiers.org/uniprot/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uniprot/': {'assoc:PairwiseGeneOrProteinInteractionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PairwiseGeneOrProteinInteractionAssociation'],\n",
       "  'input': ['http://identifiers.org/uniprot/'],\n",
       "  'api': 'Reactome API',\n",
       "  'server': 'https://reactome.org/ContentService'},\n",
       " 'http://t.biothings.io/v1/taxonomy/{taxonomyid}': {'get': {'summary': 'Retrieve taxonomy annotation based on NCBI taxon id.',\n",
       "   'parameters': [{'name': 'taxonomyid',\n",
       "     'in': 'path',\n",
       "     'description': 'NCBI Taxon ID, e.g. 9606',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/taxonomy/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/taxonomy/',\n",
       "       'template': '/taxonomy/{taxonomyid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A taxonomy object',\n",
       "     'x-responseValueType': [{'path': 'parent_taxid',\n",
       "       'valueType': 'http://identifiers.org/taxonomy/'}],\n",
       "     'x-JSONLDContext': 'taxonomy API/jsonld_context/taxonomy_1.1.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/taxonomy API/jsonld_context/taxonomy_1.1.json',\n",
       "  'output': ['http://identifiers.org/taxonomy/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/taxonomy/': {'assoc:TaxonomyIsChildOfTaxonomy'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/TaxonomyIsChildOfTaxonomy'],\n",
       "  'input': ['http://identifiers.org/taxonomy/'],\n",
       "  'api': 'Taxonomy API',\n",
       "  'server': 'http://t.biothings.io/v1'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/genesymbol/{genesymbol}/aids/JSON': {'get': {'summary': 'return pubchem bioassay ids associated with a gene symbol',\n",
       "   'parameters': [{'name': 'genesymbol',\n",
       "     'in': 'path',\n",
       "     'description': 'Put gene symbols here, e.g. CDK7 ',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/hgnc.symbol/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hgnc.symbol/',\n",
       "       'template': '/assay/target/genesymbol/{genesymbol}/aids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.AID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.bioassay/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_gene2bioassay_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_gene2bioassay_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.bioassay/': {'assoc:GeneOrGeneProductToBioassayAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToBioassayAssociation'],\n",
       "  'input': ['http://identifiers.org/hgnc.symbol/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/target/geneid/{geneid}/aids/JSON': {'get': {'summary': 'return pubchem bioassay ids associated with a gene id',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put NCBI Gene ID here, e.g. 1017',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/ncbigene/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/ncbigene/',\n",
       "       'template': '/assay/target/geneid/{geneid}/aids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.AID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.bioassay/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_gene2bioassay_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_gene2bioassay_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.bioassay/': {'assoc:GeneOrGeneProductToBioassayAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToBioassayAssociation'],\n",
       "  'input': ['http://identifiers.org/ncbigene/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/aid/{bioassayid}/sids/JSON': {'get': {'summary': 'return pubchem substance ids associated with a pubchem bioassay id',\n",
       "   'parameters': [{'name': 'bioassayid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put PubChem Bioassay ID here, e.g. 650',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubchem.bioassay/',\n",
       "       'template': '/assay/aid/{bioassayid}/sids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.SID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.substance/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_bioassay2substance_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_bioassay2substance_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.substance/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.substance/': {'assoc:BioassayToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/BioassayToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/assay/aid/{bioassayid}/cids/JSON': {'get': {'summary': 'return pubchem compound ids associated with a pubchem bioassay id',\n",
       "   'parameters': [{'name': 'bioassayid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put PubChem Bioassay ID here, e.g. 650',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubchem.bioassay/',\n",
       "       'template': '/assay/aid/{bioassayid}/cids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.CID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.compound/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_bioassay2compound_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_bioassay2compound_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.compound/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.compound/': {'assoc:BioassayToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/BioassayToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/pubchem.bioassay/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{compoundid}/sids/JSON': {'get': {'summary': 'return pubchem substance ids associated with a pubchem compound id',\n",
       "   'parameters': [{'name': 'compoundid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put PubChem Compound ID here, e.g. 2244',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/pubchem.compound/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubchem.compound/',\n",
       "       'template': '/compound/cid/{compoundid}/sids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.Information.SID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.substance/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_bioassay2substance_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_bioassay2substance_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.substance/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.substance/': {'assoc:BioassayToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/BioassayToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/pubchem.compound/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/substance/sid/{substanceid}/cids/JSON': {'get': {'summary': 'return pubchem compound ids associated with a pubchem substance id',\n",
       "   'parameters': [{'name': 'substanceid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put PubChem Substance ID here, e.g. 4594',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/pubchem.substance/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubchem.substance/',\n",
       "       'template': '/substance/sid/{substanceid}/cids/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"IdentifierList\" property',\n",
       "     'x-responseValueType': [{'path': 'IdentifierList.Information.CID',\n",
       "       'valueType': 'http://identifiers.org/pubchem.compound/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_bioassay2compound_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_bioassay2compound_context.json',\n",
       "  'output': ['http://identifiers.org/pubchem.compound/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubchem.compound/': {'assoc:BioassayToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/BioassayToChemicalAssociation'],\n",
       "  'input': ['http://identifiers.org/pubchem.substance/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{compoundid}/xrefs/PubMedID/JSON': {'get': {'summary': 'return pubmed ids associated with a pubchem compound id',\n",
       "   'parameters': [{'name': 'compoundid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put PubChem Compound ID here, e.g. 2244',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/pubchem.compound/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubchem.compound/',\n",
       "       'template': '/compound/cid/{compoundid}/xrefs/PubMedID/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"InformationList\" property',\n",
       "     'x-responseValueType': [{'path': 'InformationList.Information.PubMedID',\n",
       "       'valueType': 'http://identifiers.org/pubmed/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/pubchem_compound2pubchem_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/pubchem_compound2pubchem_context.json',\n",
       "  'output': ['http://identifiers.org/pubmed/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubmed/': {'assoc:ChemicalToPublicationAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/ChemicalToPublicationAssociation'],\n",
       "  'input': ['http://identifiers.org/pubchem.compound/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{drugname}/synonyms/JSON': {'get': {'summary': 'return synonyms of a drug name',\n",
       "   'parameters': [{'name': 'drugname',\n",
       "     'in': 'path',\n",
       "     'description': 'Put drugname here, e.g. riluzole',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "       'template': '/compound/name/{drugname}/synonyms/JSON'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A query response object with \"InformationList\" property',\n",
       "     'x-responseValueType': [{'path': 'InformationList.Information.Synonym',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'}],\n",
       "     'x-JSONLDContext': 'pubchem/jsonld_context/drugname2synonym.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pubchem/jsonld_context/drugname2synonym.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/drug-name/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'api': 'PubChem API',\n",
       "  'server': 'https://pubchem.ncbi.nlm.nih.gov/rest/pug'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/uberon/terms/{uberonid}': {'get': {'summary': 'Return uberon descriptions',\n",
       "   'parameters': [{'name': 'uberonid',\n",
       "     'in': 'path',\n",
       "     'description': 'UBERON id, e.g. 0004476',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/uberon/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uberon/',\n",
       "       'template': '/ontologies/uberon/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FUBERON_{uberonid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An UBERON description object',\n",
       "     'x-responseValueType': [{'path': 'label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/anatomy/'},\n",
       "      {'path': 'description',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/anatomy-description/'},\n",
       "      {'path': 'annotation.cross_database_reference',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/mesh.anatomy/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_term_anatomy.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_term_anatomy.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/anatomy/',\n",
       "   'http://biothings.io/explorer/vocab/terms/anatomy-description/',\n",
       "   'http://biothings.io/explorer/vocab/terms/mesh.anatomy/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/anatomy/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/mesh.anatomy/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/anatomy-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://identifiers.org/uberon/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/cl/terms/{clid}': {'get': {'summary': 'Return CL descriptions',\n",
       "   'parameters': [{'name': 'clid',\n",
       "     'in': 'path',\n",
       "     'description': 'CL id, e.g. 0000038',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/cl/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/cl/',\n",
       "       'template': '/ontologies/cl/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FCL_{clid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An CL description object',\n",
       "     'x-responseValueType': [{'path': 'label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/anatomy/'},\n",
       "      {'path': 'description',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/anatomy-description/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_term_anatomy.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_term_anatomy.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/anatomy/',\n",
       "   'http://biothings.io/explorer/vocab/terms/anatomy-description/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/anatomy/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/mesh.anatomy/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/anatomy-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://identifiers.org/cl/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/hp/terms/{hpid}': {'get': {'summary': 'Return Human Phenotypes descriptions',\n",
       "   'parameters': [{'name': 'hpid',\n",
       "     'in': 'path',\n",
       "     'description': 'HP id, e.g. 0011105',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/hp/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/hp/',\n",
       "       'template': '/ontologies/hp/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FHP_{hpid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Human Phenotypes description object',\n",
       "     'x-responseValueType': [{'path': 'label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/phenotype-name/'},\n",
       "      {'path': 'description',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/phenotype-description/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_term_phenotype.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_term_phenotype.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/phenotype-name/',\n",
       "   'http://biothings.io/explorer/vocab/terms/phenotype-description/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/phenotype-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/phenotype-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://identifiers.org/hp/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/go/terms/{goid}': {'get': {'summary': 'Return Gene Ontology descriptions',\n",
       "   'parameters': [{'name': 'goid',\n",
       "     'in': 'path',\n",
       "     'description': 'GO id, e.g. 0042535',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/go/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/go/',\n",
       "       'template': '/ontologies/go/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FGO_{goid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Gene Ontology description object',\n",
       "     'x-responseValueType': [{'path': 'description',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/go-description/'},\n",
       "      {'path': 'label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/go-term/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_term_go.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_term_go.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/go-description/',\n",
       "   'http://biothings.io/explorer/vocab/terms/go-term/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/go-term/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/go-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://identifiers.org/go/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/doid/terms/{doid}': {'get': {'summary': 'Return Disease Ontology descriptions',\n",
       "   'parameters': [{'name': 'doid',\n",
       "     'in': 'path',\n",
       "     'description': 'DO id, e.g. 126',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/doid/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/ontologies/doid/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FDOID_{doid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Disease Ontology description object',\n",
       "     'x-responseValueType': [{'path': 'description',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-description/'},\n",
       "      {'path': 'label',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_term_do.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_term_do.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/disease-description/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/disease-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/disease-description/': {'assoc:HasDescriptionAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/HasDescriptionAssociation'],\n",
       "  'input': ['http://identifiers.org/doid/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/go/children/{goid}': {'get': {'summary': 'Return children Gene Ontology id',\n",
       "   'parameters': [{'name': 'goid',\n",
       "     'in': 'path',\n",
       "     'description': 'GO id, e.g. 0042535',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/go/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/go/',\n",
       "       'template': '/ontologies/go/children?id=GO:{goid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Gene Ontology child object',\n",
       "     'x-responseValueType': [{'path': '_embedded.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/go/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_go_child.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_go_child.json',\n",
       "  'output': ['http://identifiers.org/go/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/go/': {'assoc:GeneOntologyToChildGeneOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOntologyToChildGeneOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/go/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/go/parents/{goid}': {'get': {'summary': 'Return parent Gene Ontology id',\n",
       "   'parameters': [{'name': 'goid',\n",
       "     'in': 'path',\n",
       "     'description': 'GO id, e.g. 0042535',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/go/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/go/',\n",
       "       'template': '/ontologies/go/parents?id=GO:{goid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Gene Ontology parent object',\n",
       "     'x-responseValueType': [{'path': '_embedded.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/go/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_go_parent.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_go_parent.json',\n",
       "  'output': ['http://identifiers.org/go/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/go/': {'assoc:GeneOntologyToParentGeneOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOntologyToParentGeneOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/go/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/doid/children/{doid}': {'get': {'summary': 'Return children Disease Ontology id',\n",
       "   'parameters': [{'name': 'doid',\n",
       "     'in': 'path',\n",
       "     'description': 'DO id, e.g. 162',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/doid/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/ontologies/doid/children?id=DOID:{doid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Disease Ontology child object',\n",
       "     'x-responseValueType': [{'path': '_embeddd.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/doid/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_do_child.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_do_child.json',\n",
       "  'output': ['http://identifiers.org/doid/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/doid/': {'assoc:DiseaseOntologyToChildDiseaseOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseOntologyToChildDiseaseOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/doid/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/do/parents/{doid}': {'get': {'summary': 'Return parent Disease Ontology id',\n",
       "   'parameters': [{'name': 'doid',\n",
       "     'in': 'path',\n",
       "     'description': 'DO id, e.g. 162',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/doid/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/doid/',\n",
       "       'template': '/ontologies/doid/parents?id=DOID:{doid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Disease Ontology parent object',\n",
       "     'x-responseValueType': [{'path': '_embeddd.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/doid/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_do_parent.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_do_parent.json',\n",
       "  'output': ['http://identifiers.org/doid/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/doid/': {'assoc:DiseaseOntologyToParentDiseaseOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/DiseaseOntologyToParentDiseaseOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/doid/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/uberon/parents/{uberonid}': {'get': {'summary': 'Return parent Uberon Ontology id',\n",
       "   'parameters': [{'name': 'uberonid',\n",
       "     'in': 'path',\n",
       "     'description': 'Uberon id, e.g. 0035553',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/uberon/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uberon/',\n",
       "       'template': '/ontologies/uberon/parents?id=UBERON:{uberonid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Uberon Ontology parent object',\n",
       "     'x-responseValueType': [{'path': '_embeddd.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/uberon/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_uberon_parent.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_uberon_parent.json',\n",
       "  'output': ['http://identifiers.org/uberon/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uberon/': {'assoc:AnatomyOntologyToParentAnatomyOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/AnatomyOntologyToParentAnatomyOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/uberon/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://www.ebi.ac.uk/ols/api/ontologies/uberon/children/{uberonid}': {'get': {'summary': 'Return child Uberon Ontology id',\n",
       "   'parameters': [{'name': 'uberonid',\n",
       "     'in': 'path',\n",
       "     'description': 'Uberon id, e.g. 0035553',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/uberon/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uberon/',\n",
       "       'template': '/ontologies/uberon/children?id=UBERON:{uberonid}'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'An Uberon Ontology child object',\n",
       "     'x-responseValueType': [{'path': '_embeddd.terms.obo_id',\n",
       "       'valueType': 'http://identifiers.org/uberon/'}],\n",
       "     'x-JSONLDContext': 'ebi_ontology_lookup_service_api/jsonld_context/ols_uberon_child.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ebi_ontology_lookup_service_api/jsonld_context/ols_uberon_child.json',\n",
       "  'output': ['http://identifiers.org/uberon/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/uberon/': {'assoc:AnatomyOntologyToChildAnatomyOntologyAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/AnatomyOntologyToChildAnatomyOntologyAssociation'],\n",
       "  'input': ['http://identifiers.org/uberon/'],\n",
       "  'api': 'EBI Ontology Service API',\n",
       "  'server': 'https://www.ebi.ac.uk/ols/api'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/targets({geneid})': {'get': {'summary': 'get pharos target information by pharos target id',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'Pharos target id',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/pharos.target/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.target/',\n",
       "       'template': '/targets({geneid})?view=full&format=jsonld'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos target object',\n",
       "     'x-responseValueType': [{'path': 'accession',\n",
       "       'valueType': 'http://identifiers.org/uniprot/'},\n",
       "      {'path': 'gene', 'valueType': 'http://identifiers.org/hgnc.symbol/'},\n",
       "      {'path': 'name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/gene-approved-name/'},\n",
       "      {'path': 'synonyms.Entrez Gene.term',\n",
       "       'valueType': 'http://identifiers.org/ncbigene/'},\n",
       "      {'path': 'MIM.term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.gene/'},\n",
       "      {'path': 'links.ix.idg.models.Disease.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.disease/'},\n",
       "      {'path': 'links.ix.idg.models.Disease.properties.IDG Disease/term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'},\n",
       "      {'path': 'links.ix.idg.models.Ligand.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.ligand/'},\n",
       "      {'path': 'links.ix.idg.models.Ligand.properties.IDG Ligand/term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_target.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_target.json',\n",
       "  'output': ['http://identifiers.org/uniprot/',\n",
       "   'http://identifiers.org/hgnc.symbol/',\n",
       "   'http://biothings.io/explorer/vocab/terms/gene-approved-name/',\n",
       "   'http://identifiers.org/ncbigene/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.gene/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.ligand/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hgnc.symbol/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/uniprot/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/gene-approved-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/ncbigene/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/hgnc/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.disease/': {'assoc:GeneOrGeneProductToDiseaseAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.ligand/': {'assoc:GeneOrGeneProductToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToChemicalAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/GeneOrGeneProductToDiseaseAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/pharos.target/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/ligands({drugid})': {'get': {'summary': 'get pharos ligand information by pharos ligand id',\n",
       "   'parameters': [{'name': 'drugid',\n",
       "     'in': 'path',\n",
       "     'description': 'Pharos ligand id',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/pharos.ligand/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.ligand/',\n",
       "       'template': '/ligands({drugid})?view=full&format=jsonld'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos ligand object',\n",
       "     'x-responseValueType': [{'path': 'name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'},\n",
       "      {'path': 'synonyms.CHEMBL ID',\n",
       "       'valueType': 'http://identifiers.org/chembl.compound/'},\n",
       "      {'path': 'links.ix.idg.models.Target.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.target/'},\n",
       "      {'path': 'links.ix.idg.models.Target.properties.IDG Target.term',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'},\n",
       "      {'path': 'links.ix.idg.models.Disease.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.disease/'},\n",
       "      {'path': 'links.ix.idg.models.Disease.properties.IDG Disease/term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_ligand.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_ligand.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "   'http://identifiers.org/chembl.compound/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.target/',\n",
       "   'http://identifiers.org/hgnc.symbol/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/drug-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/chembl.compound/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.disease/': {'assoc:ChemicalToDiseaseOrPhenotypicFeatureAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.target/': {'assoc:ChemicalToGeneOrGeneProductAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/ChemicalToGeneOrGeneProductAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/ChemicalToDiseaseOrPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/pharos.ligand/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/diseases({diseaseid})': {'get': {'summary': 'get pharos disease information by pharos disease id',\n",
       "   'parameters': [{'name': 'diseaseid',\n",
       "     'in': 'path',\n",
       "     'description': 'Pharos disease id',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/pharos.disease/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.disease/',\n",
       "       'template': '/diseases({diseaseid})?view=full&format=jsonld'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos ligand object',\n",
       "     'x-responseValueType': [{'path': 'name',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/'},\n",
       "      {'path': 'synonyms.DOID', 'valueType': 'http://identifiers.org/doid/'},\n",
       "      {'path': 'synonyms.DiseaseOntology.term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/omim.disease/'},\n",
       "      {'path': 'synonyms.DiseaseOntology.term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/umls.disease/'},\n",
       "      {'path': 'links.ix.idg.models.Target.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.target/'},\n",
       "      {'path': 'links.ix.idg.models.Target.properties.IDG Target.term',\n",
       "       'valueType': 'http://identifiers.org/hgnc.symbol/'},\n",
       "      {'path': 'links.ix.idg.models.Ligand.refid',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.ligand/'},\n",
       "      {'path': 'links.ix.idg.models.Ligand.properties.IDG Ligand/term',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_disease.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_disease.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "   'http://identifiers.org/doid/',\n",
       "   'http://biothings.io/explorer/vocab/terms/omim.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/umls.disease/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.target/',\n",
       "   'http://identifiers.org/hgnc.symbol/',\n",
       "   'http://biothings.io/explorer/vocab/terms/pharos.ligand/',\n",
       "   'http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/disease-name/': {'assoc:EquivalentAssociation'},\n",
       "               'http://identifiers.org/doid/': {'assoc:EquivalentAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.target/': {'assoc:DiseaseToGeneOrGeneProductAssociation'},\n",
       "               'http://biothings.io/explorer/vocab/terms/pharos.disease/': {'assoc:DiseaseToChemicalAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/DiseaseToChemicalAssociation',\n",
       "   'http://biothings.io/explorer/vocab/objects/DiseaseToGeneOrGeneProductAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/pharos.disease/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/ligands/search': {'get': {'summary': 'return pharos ligand id based on drugname',\n",
       "   'parameters': [{'name': 'q',\n",
       "     'in': 'query',\n",
       "     'description': 'Put drugname here, e.g. riluzole ',\n",
       "     'required': True,\n",
       "     'example': 'riluzole',\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/drug-name/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos ligand annotation object',\n",
       "     'x-responseValueType': [{'path': 'content.id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.ligand/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_drugname2ligand.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_drugname2ligand.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/pharos.ligand/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/pharos.ligand/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/drug-name/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/targets/search': {'get': {'summary': 'return pharos target id based on uniprot',\n",
       "   'parameters': [{'name': 'q',\n",
       "     'in': 'query',\n",
       "     'description': 'Put uniprot id here, e.g. P61073',\n",
       "     'required': True,\n",
       "     'example': 'P61073',\n",
       "     'x-valueType': ['http://identifiers.org/uniprot/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/uniprot/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos target annotation object',\n",
       "     'x-responseValueType': [{'path': 'content.id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.target/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_uniprot2target.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_uniprot2target.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/pharos.target/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/pharos.target/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://identifiers.org/uniprot/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://pharos.nih.gov/idg/api/v1/diseases/search': {'get': {'summary': 'return pharos disease id based on disease name',\n",
       "   'parameters': [{'name': 'q',\n",
       "     'in': 'query',\n",
       "     'description': 'Put disease name here, e.g. cancer',\n",
       "     'required': True,\n",
       "     'example': 'cancer',\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/disease-name/',\n",
       "       'template': 'input_value'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A pharos disease annotation object',\n",
       "     'x-responseValueType': [{'path': 'content.id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/pharos.disease/'}],\n",
       "     'x-JSONLDContext': 'pharos/jsonld_context/pharos_diseasename2disease.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/pharos/jsonld_context/pharos_diseasename2disease.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/pharos.disease/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://biothings.io/explorer/vocab/terms/pharos.disease/': {'assoc:EquivalentAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/EquivalentAssociation'],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/disease-name/'],\n",
       "  'api': 'Pharos API',\n",
       "  'server': 'https://pharos.nih.gov/idg/api/v1'},\n",
       " 'https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1/public/rsids2pmids': {'get': {'summary': 'Retrieve PMIDs of publications mentioning submitted variants',\n",
       "   'parameters': [{'name': 'rsids',\n",
       "     'in': 'query',\n",
       "     'description': 'rsid',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/dbsnp/'],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'publications mentioned the submitted variant',\n",
       "     'x-responseValueType': [{'path': 'data.pmids',\n",
       "       'valueType': 'http://identifiers.org/pubmed/'}],\n",
       "     'x-JSONLDContext': 'litvar/jsonld_context/rsid2pmid_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/litvar/jsonld_context/rsid2pmid_context.json',\n",
       "  'output': ['http://identifiers.org/pubmed/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/pubmed/': {'assoc:VariantToPublicationAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/VariantToPublicationAssociation'],\n",
       "  'input': ['http://identifiers.org/dbsnp/'],\n",
       "  'api': 'LitVar API',\n",
       "  'server': 'https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1'},\n",
       " 'http://rest.ensembl.org/variation/human/pmid/{pmid}': {'get': {'summary': 'Fetch variants by publication using PubMed reference number (PMID)',\n",
       "   'parameters': [{'name': 'pmid',\n",
       "     'in': 'path',\n",
       "     'description': 'PubMed reference number (PMID)',\n",
       "     'required': True,\n",
       "     'example': 26318936,\n",
       "     'x-valueType': ['http://identifiers.org/pubmed/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/pubmed/',\n",
       "       'template': '/variation/human/pmid/{pmid}?content-type=application/json'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A variant annotation object',\n",
       "     'x-responseValueType': [{'path': 'name',\n",
       "       'valueType': 'http://identifiers.org/dbsnp/'}],\n",
       "     'x-JSONLDContext': 'ensembl/jsonld_context/pmid2rsid_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ensembl/jsonld_context/pmid2rsid_context.json',\n",
       "  'output': ['http://identifiers.org/dbsnp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/dbsnp/': {'assoc:PublicationToVariantAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/PublicationToVariantAssociation'],\n",
       "  'input': ['http://identifiers.org/pubmed/'],\n",
       "  'api': 'Ensembl API',\n",
       "  'server': 'http://rest.ensembl.org'},\n",
       " 'http://rest.ensembl.org/variation/human/{rsid}': {'get': {'summary': 'Uses a variant identifier (e.g. rsID) to return phenotypic feature associated with the variant',\n",
       "   'parameters': [{'name': 'rsid',\n",
       "     'in': 'path',\n",
       "     'description': 'Variant id',\n",
       "     'required': True,\n",
       "     'example': 'rs56116432',\n",
       "     'x-valueType': ['http://identifiers.org/dbsnp/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/dbsnp/',\n",
       "       'template': '/variation/human/{rsid}?phenotypes=1;content-type=application/json'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'A phenoytpe annotation object',\n",
       "     'x-responseValueType': [{'path': 'phenotypes.trait',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/phenotype-name/'},\n",
       "      {'path': 'phenotypes.ontology_accessions',\n",
       "       'valueType': 'http://identifiers.org/hp/'}],\n",
       "     'x-JSONLDContext': 'ensembl/jsonld_context/rsid2phenotype_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ensembl/jsonld_context/rsid2phenotype_context.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/phenotype-name/',\n",
       "   'http://identifiers.org/hp/'],\n",
       "  'relation': defaultdict(set,\n",
       "              {'http://identifiers.org/hp/': {'assoc:VariantToDiseaseOrPhenotypicFeatureAssociation'}}),\n",
       "  'associations': ['http://biothings.io/explorer/vocab/objects/VariantToDiseaseOrPhenotypicFeatureAssociation'],\n",
       "  'input': ['http://identifiers.org/dbsnp/'],\n",
       "  'api': 'Ensembl API',\n",
       "  'server': 'http://rest.ensembl.org'},\n",
       " 'http://rest.ensembl.org/taxonomy/id/{taxonomyid}': {'get': {'summary': 'Search for a taxonomic term by its identifier or name',\n",
       "   'parameters': [{'name': 'taxonomyid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put taxonomy ids here, e.g. 9606',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://identifiers.org/taxonomy/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://identifiers.org/taxonomy/',\n",
       "       'template': '/taxonomy/id/{taxonomyid}?content-type=application/json'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'taxonomy annotation information',\n",
       "     'x-responseValueType': [{'path': 'parent.id',\n",
       "       'valueType': 'http://identifiers.org/taxonomy/'}],\n",
       "     'x-JSONLDContext': 'ensembl/jsonld_context/ensembl_taxonomy_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ensembl/jsonld_context/ensembl_taxonomy_context.json',\n",
       "  'output': ['http://identifiers.org/taxonomy/'],\n",
       "  'relation': defaultdict(set, {}),\n",
       "  'associations': [],\n",
       "  'input': ['http://identifiers.org/taxonomy/'],\n",
       "  'api': 'Ensembl API',\n",
       "  'server': 'http://rest.ensembl.org'},\n",
       " 'http://rest.ensembl.org/lookup/id/{geneid}': {'get': {'summary': 'Find the species and database for a single identifier e.g. gene, transcript, protein',\n",
       "   'parameters': [{'name': 'geneid',\n",
       "     'in': 'path',\n",
       "     'description': 'Put ensembl gene ids here, e.g. ENSG00000157764',\n",
       "     'required': True,\n",
       "     'x-valueType': ['http://biothings.io/explorer/vocab/terms/ensembl.gene/'],\n",
       "     'x-requestTemplate': [{'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.gene/',\n",
       "       'template': '/lookup/id/{geneid}?content-type=application/json;phenotypes=1;expand=1'}],\n",
       "     'schema': {'type': 'string'}}],\n",
       "   'responses': {'200': {'description': 'ensembl gene lookup information',\n",
       "     'x-responseValueType': [{'path': 'Transcript.Exon.id',\n",
       "       'valueType': 'http://identifiers.org/ensembl.exon/'},\n",
       "      {'path': 'Transcript.Translation.id',\n",
       "       'valueType': 'http://biothings.io/explorer/vocab/terms/ensembl.protein/'},\n",
       "      {'path': 'phenotypes.ontology_accessions',\n",
       "       'valueType': 'http://identifiers.org/efo/'},\n",
       "      {'path': 'phenotypes.ontology_accessions',\n",
       "       'valueType': 'http://identifiers.org/hp/'}],\n",
       "     'x-JSONLDContext': 'ensembl/jsonld_context/ensembl_gene_context.json'}}},\n",
       "  'jsonld_context': '/Users/kevinxin/Documents/work/biothings_explorer/biothings_explorer_web/src/biothings_explorer/openapi_specs/ensembl/jsonld_context/ensembl_gene_context.json',\n",
       "  'output': ['http://biothings.io/explorer/vocab/terms/ensembl.protein/',\n",
       "   'http://identifiers.org/efo/',\n",
       "   'http://identifiers.org/hp/'],\n",
       "  'relation': defaultdict(set, {}),\n",
       "  'associations': [],\n",
       "  'input': ['http://biothings.io/explorer/vocab/terms/ensembl.gene/'],\n",
       "  'api': 'Ensembl API',\n",
       "  'server': 'http://rest.ensembl.org'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registry.endpoint_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'http://identifiers.org/uberon/': {'description': 'Uberon is an integrated cross-species anatomy ontology representing a variety of entities classified according to traditional anatomical criteria such as structure, function and developmental lineage. The ontology includes comprehensive relationships to taxon-specific anatomical ontologies, allowing integration of functional, phenotype and expression data.',\n",
       "  'preferred_name': 'UBERON',\n",
       "  'semantic type': 'anatomy',\n",
       "  'pattern': '^UBERON\\\\:\\\\d+$',\n",
       "  'prefix': 'uberon',\n",
       "  'example': 'UBERON:0008203',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/mesh.anatomy/': {'description': \"MeSH (Medical Subject Headings) is the National Library of Medicine's controlled vocabulary thesaurus. It consists of sets of terms naming descriptors in a hierarchical structure that permits searching at various levels of specificity. This thesaurus is used by NLM for indexing articles from biomedical journals, cataloguing of books, docume\",\n",
       "  'preferred_name': 'mesh.anatomy',\n",
       "  'semantic type': 'anatomy',\n",
       "  'pattern': None,\n",
       "  'prefix': 'mesh.anatomy',\n",
       "  'example': '9738',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/cl/': {'description': 'The Cell Ontology is designed as a structured controlled vocabulary for cell types. The ontology was constructed for use by the model organism and other bioinformatics databases, incorporating cell types from prokaryotes to mammals, and includes plants and fungi.',\n",
       "  'preferred_name': 'Cell Type Ontology',\n",
       "  'semantic type': 'anatomy',\n",
       "  'pattern': '^CL:\\\\d{7}$',\n",
       "  'prefix': 'cl',\n",
       "  'example': 'CL:0000232',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/anatomy-description/': {'description': 'description of anatomy',\n",
       "  'preferred_name': 'Anatomy-description',\n",
       "  'semantic type': 'anatomy',\n",
       "  'pattern': None,\n",
       "  'prefix': 'anatomy-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/anatomy/': {'description': 'anatomy name',\n",
       "  'preferred_name': 'anatomy',\n",
       "  'semantic type': 'anatomy',\n",
       "  'pattern': None,\n",
       "  'prefix': 'anatomy',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://identifiers.org/pubchem.bioassay/': {'description': \"PubChem provides information on the biological activities of small molecules. It is a component of NIH's Molecular Libraries Roadmap Initiative. PubChem bioassay archives active compounds and bioassay results.\",\n",
       "  'preferred_name': 'PubChem-bioassay',\n",
       "  'semantic type': 'bioassay',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pubchem.bioassay',\n",
       "  'example': '1018',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/chembl.compound/': {'description': 'ChEMBL is a database of bioactive compounds, their quantitative properties and bioactivities (binding constants, pharmacology and ADMET, etc). The data is abstracted and curated from the primary scientific literature.',\n",
       "  'preferred_name': 'ChEMBL compound',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^CHEMBL\\\\d+$',\n",
       "  'prefix': 'chembl.compound',\n",
       "  'example': 'CHEMBL744',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/drug-name/': {'description': 'name of a drug',\n",
       "  'preferred_name': 'drug-name',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'drug-name',\n",
       "  'example': 'riluzole',\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/chemical-description/': {'description': 'description of a drug',\n",
       "  'preferred_name': 'chemical-description',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'chemical-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://identifiers.org/inchikey/': {'description': 'The IUPAC International Chemical Identifier (InChI, see MIR:00000383) is an identifier for chemical substances, and is derived solely from a structural representation of that substance. Since these can be quite unwieldly, particularly for web use, the InChIKey was developed. These are of a fixed length (25 character) and were created as a condensed, more web friendly, digital representation of the InChI.',\n",
       "  'preferred_name': 'InChIKey',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^[A-Z]{14}\\\\-[A-Z]{10}(\\\\-[A-Z])?',\n",
       "  'prefix': 'inchikey',\n",
       "  'example': 'RYYVLZVUVIJVGH-UHFFFAOYSA-N',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/pharos.ligand/': {'description': 'pharos ligand id',\n",
       "  'preferred_name': 'pharos.ligand',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pharos.ligand',\n",
       "  'example': '1',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/rxcui/': {'description': 'The RxNorm concept unique identifer for the clinical drug or substance involved in the event',\n",
       "  'preferred_name': 'rxcui',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'rxcui',\n",
       "  'example': '35623',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/chebi/': {'description': \"Chemical Entities of Biological Interest (ChEBI) is a freely available dictionary of molecular entities focused on 'small' chemical compounds.\",\n",
       "  'preferred_name': 'ChEBI',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^CHEBI:\\\\d+$',\n",
       "  'prefix': 'chebi',\n",
       "  'example': 'CHEBI:36927',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/umls.compound/': {'description': 'The Unified Medical Language System is a repository of biomedical vocabularies. Vocabularies integrated in the UMLS Metathesaurus include the NCBI taxonomy, Gene Ontology, the Medical Subject Headings (MeSH), OMIM and the Digital Anatomist Symbolic Knowledge Base. UMLS concepts are not only inter-related, but may also be linked to external resources such as GenBank.',\n",
       "  'preferred_name': 'umls.compound',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^C\\\\d+$',\n",
       "  'prefix': 'umls.compound',\n",
       "  'example': 'C0073379',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/mesh.compound/': {'description': 'mesh compound id',\n",
       "  'preferred_name': 'mesh.compound',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^(C|D)\\\\d{6}$',\n",
       "  'prefix': 'mesh.compound',\n",
       "  'example': 'D019782',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/drugbank/': {'description': 'The DrugBank database is a bioinformatics and chemoinformatics resource that combines detailed drug (i.e. chemical, pharmacological and pharmaceutical) data with comprehensive drug target (i.e. sequence, structure, and pathway) information. This collection references drug information.',\n",
       "  'preferred_name': 'DrugBank',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^DB\\\\d{5}$',\n",
       "  'prefix': 'drugbank',\n",
       "  'example': 'DB00001',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/iuphar.ligand/': {'description': 'The IUPHAR Compendium details the molecular, biophysical and pharmacological properties of identified mammalian sodium, calcium and potassium channels, as well as the related cyclic nucleotide-modulated ion channels and the recently described transient receptor potential channels. It includes information on nomenclature systems, and on inter and intra-species molecular structure variation. This collection references ligands.',\n",
       "  'preferred_name': 'IUPHAR ligand',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'iuphar.ligand',\n",
       "  'example': '1755',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/kegg.drug/': {'description': 'KEGG DRUG contains chemical structures of drugs and additional information such as therapeutic categories and target molecules.',\n",
       "  'preferred_name': 'KEGG Drug',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^D\\\\d+$',\n",
       "  'prefix': 'kegg.drug',\n",
       "  'example': 'D00123',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/kegg.compound/': {'description': 'KEGG compound contains our knowledge on the universe of chemical substances that are relevant to life.',\n",
       "  'preferred_name': 'KEGG Compound',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^C\\\\d+$',\n",
       "  'prefix': 'kegg.compound',\n",
       "  'example': 'C12345',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pubchem.compound/': {'description': \"PubChem provides information on the biological activities of small molecules. It is a component of NIH's Molecular Libraries Roadmap Initiative. PubChem Compound archives chemical structures and records.\",\n",
       "  'preferred_name': 'PubChem-compound',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pubchem.compound',\n",
       "  'example': '5070',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pubchem.substance/': {'description': \"PubChem provides information on the biological activities of small molecules. It is a component of NIH's Molecular Libraries Roadmap Initiative. PubChem Substance archives chemical substance records.\",\n",
       "  'preferred_name': 'PubChem-substance',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pubchem.substance',\n",
       "  'example': '223440865',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/unii/': {'description': 'The purpose of the joint FDA/USP Substance Registration System (SRS) is to support health information technology initiatives by generating unique ingredient identifiers (UNIIs) for substances in drugs, biologics, foods, and devices. The UNII is a non- proprietary, free, unique, unambiguous, non semantic, alphanumeric identifier based on a substance\\x89s molecular structure and/or descriptive information.',\n",
       "  'preferred_name': 'UNII',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^[A-Z0-9]+$',\n",
       "  'prefix': 'unii',\n",
       "  'example': '3G6A5W338E',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/inchi/': {'description': 'The IUPAC International Chemical Identifier (InChI) is a non-proprietary identifier for chemical substances that can be used in printed and electronic data sources. It is derived solely from a structural representation of that substance, such that a single compound always yields the same identifier.',\n",
       "  'preferred_name': 'InChI',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^InChI\\\\=1S?\\\\/[A-Za-z0-9\\\\.]+(\\\\+[0-9]+)?(\\\\/[cnpqbtmsih][A-Za-z0-9\\\\-\\\\+\\\\(\\\\)\\\\,\\\\/\\\\?\\\\;\\\\.]+)*$',\n",
       "  'prefix': 'inchi',\n",
       "  'example': 'InChI=1S/C2H6O/c1-2-3/h3H,2H2,1H3',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pharmgkb.drug/': {'description': 'The PharmGKB database is a central repository for genetic, genomic, molecular and cellular phenotype data and clinical information about people who have participated in pharmacogenomics research studies. The data includes, but is not limited to, clinical and basic pharmacokinetic and pharmacogenomic research in the cardiovascular, pulmonary, cancer, pathways, metabolic and transporter domains.',\n",
       "  'preferred_name': 'PharmGKB Drug',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': '^PA\\\\d+$',\n",
       "  'prefix': 'pharmgkb.drug',\n",
       "  'example': 'PA448710',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/ndfrt/': {'description': 'NDF-RT combines the NDF hierarchical drug classification with a multi-category reference model.',\n",
       "  'preferred_name': 'ndfrt',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'ndfrt',\n",
       "  'example': 'N0000006860',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/epc-description/': {'description': 'An FDA Established Pharmacologic Class (EPC) text phrase is a pharmacologic class associated with an approved indication of an active moiety that the FDA has determined to be scientifically valid and clinically meaningful.',\n",
       "  'preferred_name': 'EPC-Description',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'epc-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/drug-phase/': {'description': 'drug phase',\n",
       "  'preferred_name': 'Drug-Phase',\n",
       "  'semantic type': 'chemical',\n",
       "  'pattern': None,\n",
       "  'prefix': 'drug-phase',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://identifiers.org/snomedct/': {'description': 'SNOMED CT (Systematized Nomenclature of Medicine -- Clinical Terms), is a systematically organized computer processable collection of medical terminology covering most areas of clinical information such as diseases, findings, procedures, microorganisms, pharmaceuticals, etc.',\n",
       "  'preferred_name': 'SNOMED CT',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^(\\\\w+)?\\\\d+$',\n",
       "  'prefix': 'snomedct',\n",
       "  'example': '284196006',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/disease-name/': {'description': 'disease name',\n",
       "  'preferred_name': 'disease-name',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': None,\n",
       "  'prefix': 'disease-name',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://identifiers.org/meddra/': {'description': 'The Medical Dictionary for Regulatory Activities (MedDRA) was developed by the International Council for Harmonisation of Technical Requirements for Registration of Pharmaceuticals for Human Use (ICH)to provide a standardised medical terminology to facilitate sharing of regulatory information internationally for medical products used by humans. It is used within regulatory processes, safety monitoring, as well as for marketing activities. Products covered by the scope of MedDRA include pharmaceuticals, biologics, vaccines and drug-device combination products. The MedDRA dictionary is organized by System Organ Class (SOC), divided into High-Level Group Terms (HLGT), High-Level Terms (HLT), Preferred Terms (PT) and finally into Lowest Level Terms (LLT).',\n",
       "  'preferred_name': 'MedDRA',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'meddra',\n",
       "  'example': '10015919',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/orphanet/': {'description': 'Orphanet is a reference portal for information on rare diseases and orphan drugs. It\\x89s aim is to help improve the diagnosis, care and treatment of patients with rare diseases.',\n",
       "  'preferred_name': 'Orphanet',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'orphanet',\n",
       "  'example': '85163',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/disease-description/': {'description': 'disease description',\n",
       "  'preferred_name': 'disease-description',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': None,\n",
       "  'prefix': 'disease-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/mesh.disease/': {'description': \"MeSH (Medical Subject Headings) is the National Library of Medicine's controlled vocabulary thesaurus. It consists of sets of terms naming descriptors in a hierarchical structure that permits searching at various levels of specificity. This thesaurus is used by NLM for indexing articles from biomedical journals, cataloguing of books, docume\",\n",
       "  'preferred_name': 'mesh.disease',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^(C|D)\\\\d{6}$',\n",
       "  'prefix': 'mesh.disease',\n",
       "  'example': 'D007619',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/ncit/': {'description': 'NCI Thesaurus (NCIt) provides reference terminology covering vocabulary for clinical care, translational and basic research, and public information and administrative activities, providing a stable and unique identification code.',\n",
       "  'preferred_name': 'NCIt',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^C\\\\d+$',\n",
       "  'prefix': 'ncit',\n",
       "  'example': 'C80519',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/pharos.disease/': {'description': 'pharos disease id',\n",
       "  'preferred_name': 'pharos.disease',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pharos.disease',\n",
       "  'example': '1',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/umls.disease/': {'description': 'The Unified Medical Language System is a repository of biomedical vocabularies. Vocabularies integrated in the UMLS Metathesaurus include the NCBI taxonomy, Gene Ontology, the Medical Subject Headings (MeSH), OMIM and the Digital Anatomist Symbolic Knowledge Base. UMLS concepts are not only inter-related, but may also be linked to external resources such as GenBank.',\n",
       "  'preferred_name': 'umls.disease',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^C\\\\d+$',\n",
       "  'prefix': 'umls.disease',\n",
       "  'example': 'C0022521',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/omim.disease/': {'description': 'omim disease id',\n",
       "  'preferred_name': 'omim.disease',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^[*#+%^]?\\\\d{6}$',\n",
       "  'prefix': 'omim.disease',\n",
       "  'example': '244400',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/mondo/': {'description': 'mondo disease id',\n",
       "  'preferred_name': 'MONDO',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^MONDO\\\\:\\\\d+$',\n",
       "  'prefix': 'mondo',\n",
       "  'example': 'MONDO:0016575',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/doid/': {'description': 'The Disease Ontology has been developed as a standardized ontology for human disease with the purpose of providing the biomedical community with consistent, reusable and sustainable descriptions of human disease terms, phenotype characteristics and related medical vocabulary disease concepts.',\n",
       "  'preferred_name': 'Human Disease Ontology',\n",
       "  'semantic type': 'disease',\n",
       "  'pattern': '^DOID\\\\:\\\\d+$',\n",
       "  'prefix': 'doid',\n",
       "  'example': 'DOID:11337',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/hgnc/': {'description': 'The HGNC (HUGO Gene Nomenclature Committee) provides an approved gene name and symbol (short-form abbreviation) for each known human gene.  All approved symbols are stored in the HGNC database, and each symbol is unique. HGNC identifiers refer to records in the HGNC symbol database.',\n",
       "  'preferred_name': 'HGNC',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^((HGNC|hgnc):)?\\\\d{1,5}$',\n",
       "  'prefix': 'hgnc',\n",
       "  'example': '2674',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/hgnc.symbol/': {'description': 'The HGNC (HUGO Gene Nomenclature Committee) provides an approved gene name and symbol (short-form abbreviation) for each known human gene. All approved symbols are stored in the HGNC database, and each symbol is unique. This collection refers to records using the HGNC symbol.',\n",
       "  'preferred_name': 'HGNC Symbol',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^[A-Za-z-0-9_]+(\\\\@)?$',\n",
       "  'prefix': 'hgnc.symbol',\n",
       "  'example': 'CDK7',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/ncbigene/': {'description': \"Entrez Gene is the NCBI's database for gene-specific information, focusing on completely sequenced genomes, those with an active research community to contribute gene-specific information, or those that are scheduled for intense sequence analysis.\",\n",
       "  'preferred_name': 'NCBI Gene',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'ncbigene',\n",
       "  'example': '1019',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/umls.gene/': {'description': 'umls gene id',\n",
       "  'preferred_name': 'umls.gene',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^C\\\\d+$',\n",
       "  'prefix': 'umls.gene',\n",
       "  'example': 'C1332733',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/gene-approved-name/': {'description': 'gene approved name',\n",
       "  'preferred_name': 'gene-approved-name',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': None,\n",
       "  'prefix': 'gene-approved-name',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/gene-description/': {'description': 'description for gene',\n",
       "  'preferred_name': 'gene-description',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': None,\n",
       "  'prefix': 'gene-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://identifiers.org/ccds/': {'description': 'The Consensus CDS (CCDS) project is a collaborative effort to identify a core set of human and mouse protein coding regions that are consistently annotated and of high quality. The CCDS set is calculated following coordinated whole genome annotation updates carried out by the NCBI, WTSI, and Ensembl. The long term goal is to support convergence towards a standard set of gene annotations.',\n",
       "  'preferred_name': 'Consensus CDS',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^CCDS\\\\d+\\\\.\\\\d+$',\n",
       "  'prefix': 'ccds',\n",
       "  'example': 'CCDS13573.1',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/ena.embl/': {'description': 'The European Nucleotide Archive (ENA) captures and presents information relating to experimental workflows that are based around nucleotide sequencing. ENA is made up of a number of distinct databases that includes EMBL-Bank, the Sequence Read Archive (SRA) and the Trace Archive each with their own data formats and standards. This collection references Embl-Bank identifiers.',\n",
       "  'preferred_name': 'ENA',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^[A-Z]+[0-9]+(\\\\.\\\\d+)?$',\n",
       "  'prefix': 'ena.embl',\n",
       "  'example': 'BN000065',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/ensembl.gene/': {'description': 'ensembl gene id',\n",
       "  'preferred_name': 'ensembl.gene',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^((ENS[FPTG]\\\\\\\\d{11}(\\\\\\\\.\\\\\\\\d+)?)|(FB\\\\\\\\w{2}\\\\\\\\d{7})|(Y[A-Z]{2}\\\\\\\\d{3}[a-zA-Z](\\\\\\\\-[A-Z])?)|([A-Z_a-z0-9]+(\\\\\\\\.)?(t)?(\\\\\\\\d+)?([a-z])?))$',\n",
       "  'prefix': 'ensembl.gene',\n",
       "  'example': 'ENSG00000123374',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/mirbase/': {'description': 'The miRBase Sequence Database is a searchable database of published miRNA sequences and annotation. The data were previously provided by the miRNA Registry. Each entry in the miRBase Sequence database represents a predicted hairpin portion of a miRNA transcript (termed mir in the database), with information on the location and sequence of the mature miRNA sequence (termed miR).',\n",
       "  'preferred_name': 'miRBase Sequence',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': 'MI\\\\d{7}',\n",
       "  'prefix': 'mirbase',\n",
       "  'example': 'MI0000001',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/mgi/': {'description': 'The Mouse Genome Database (MGD) project includes data on gene characterization, nomenclature, mapping, gene homologies among mammals, sequence links, phenotypes, allelic variants and mutants, and strain data.',\n",
       "  'preferred_name': 'Mouse Genome Database',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^MGI:\\\\d+$',\n",
       "  'prefix': 'mgi',\n",
       "  'example': 'MGI:2442292',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/wormbase/': {'description': 'The Mouse Genome Database (MGD) project includes data on gene characterization, nomenclature, mapping, gene homologies among mammals, sequence links, phenotypes, allelic variants and mutants, and strain data.',\n",
       "  'preferred_name': 'Worm Base',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^Worm:\\\\d+$',\n",
       "  'prefix': 'mgi',\n",
       "  'example': 'WBGene00013782',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/fb/': {'description': 'flybase id',\n",
       "  'preferred_name': 'flybase',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^FB\\\\w{2}\\\\d{7}$',\n",
       "  'prefix': 'flybase',\n",
       "  'example': 'FBgn0011293',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/xenbase/': {'description': 'Xenbase is the model organism database for Xenopus laevis and X. (Silurana) tropicalis. It contains genomic, development data and community information for Xenopus research. it includes gene expression patterns that incorporates image data from the literature, large scale screens and community submissions.',\n",
       "  'preferred_name': 'Xenbase',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^(XB-GENE-)?\\\\d+$',\n",
       "  'prefix': 'xenbase',\n",
       "  'example': '922462',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/omim.gene/': {'description': 'omim gene id',\n",
       "  'preferred_name': 'omim.gene',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^[*#+%^]?\\\\d{6}$',\n",
       "  'prefix': 'omim.gene',\n",
       "  'example': '116953',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/refseq/': {'description': 'The Reference Sequence (RefSeq) collection aims to provide a comprehensive, integrated, non-redundant set of sequences, including genomic DNA, transcript (RNA), and protein products.',\n",
       "  'preferred_name': 'RefSeq',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^((AC|AP|NC|NG|NM|NP|NR|NT|NW|XM|XP|XR|YP|ZP)_\\\\d+|(NZ\\\\_[A-Z]{4}\\\\d+))(\\\\.\\\\d+)?$',\n",
       "  'prefix': 'refseq',\n",
       "  'example': 'NP_012345',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/rgd/': {'description': 'Rat Genome Database seeks to collect, consolidate, and integrate rat genomic and genetic data with curated functional and physiological data and make these data widely available to the scientific community. This collection references genes.',\n",
       "  'preferred_name': 'Rat Genome Database',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^\\\\d{4,}$',\n",
       "  'prefix': 'rgd',\n",
       "  'example': '2018',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/sgd/': {'description': 'The Saccharomyces Genome Database (SGD) project collects information and maintains a database of the molecular biology of the yeast Saccharomyces cerevisiae.',\n",
       "  'preferred_name': 'SGD',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^((S\\\\d+$)|(Y[A-Z]{2}\\\\d{3}[a-zA-Z](\\\\-[A-Z])?))$',\n",
       "  'prefix': 'sgd',\n",
       "  'example': 'S000003909',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/zfin/': {'description': 'ZFIN serves as the zebrafish model organism database. This collection references all zebrafish biological entities in ZFIN.',\n",
       "  'preferred_name': 'ZFIN Bioentity',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^ZDB\\\\-\\\\w+\\\\-\\\\d+\\\\-\\\\d+$',\n",
       "  'prefix': 'zfin',\n",
       "  'example': 'ZDB-GENE-041118-11',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/pharos.target/': {'description': 'pharos target id',\n",
       "  'preferred_name': 'pharos.target',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pharos.target',\n",
       "  'example': '1',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/reactome.complex/': {'description': 'reacomte complex id',\n",
       "  'preferred_name': 'reactome.complex',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '(^R-[A-Z]{3}-\\\\d+(-\\\\d+)?(\\\\.\\\\d+)?$)|(^REACT_\\\\d+(\\\\.\\\\d+)?$)',\n",
       "  'prefix': 'reactome.complex',\n",
       "  'example': 'R-HSA-1497822',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/ro/': {'description': 'The OBO Relation Ontology provides consistent and unambiguous formal definitions of the relational expressions used in biomedical ontologies.',\n",
       "  'preferred_name': 'Relation Ontology',\n",
       "  'semantic type': 'ontology',\n",
       "  'pattern': '^RO:\\\\d+{7}$',\n",
       "  'prefix': 'ro',\n",
       "  'example': 'RO:0002533',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/go/': {'description': 'The Gene Ontology project provides a controlled vocabulary to describe gene and gene product attributes in any organism.',\n",
       "  'preferred_name': 'Gene Ontology',\n",
       "  'semantic type': 'ontology',\n",
       "  'pattern': '^GO:\\\\d{7}$',\n",
       "  'prefix': 'go',\n",
       "  'example': 'GO:0006915',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/go-term/': {'description': 'go term label',\n",
       "  'preferred_name': 'go-term',\n",
       "  'semantic type': 'ontology',\n",
       "  'pattern': None,\n",
       "  'prefix': 'go-term',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/go-description/': {'description': 'description for go term',\n",
       "  'preferred_name': 'godescription',\n",
       "  'semantic type': 'ontology',\n",
       "  'pattern': None,\n",
       "  'prefix': 'godescription',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/go-evidence/': {'description': 'go evidence code',\n",
       "  'preferred_name': 'go-evidence',\n",
       "  'semantic type': 'ontology',\n",
       "  'pattern': None,\n",
       "  'prefix': 'go-evidence',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://identifiers.org/taxonomy/': {'description': 'The taxonomy contains the relationships between all living forms for which nucleic acid or protein sequence have been determined.',\n",
       "  'preferred_name': 'Taxonomy',\n",
       "  'semantic type': 'organism',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'taxonomy',\n",
       "  'example': '9606',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/taxonomyname/': {'description': 'taxonomy label',\n",
       "  'preferred_name': 'taxonomyname',\n",
       "  'semantic type': 'taxonomy',\n",
       "  'pattern': None,\n",
       "  'prefix': 'taxonomyname',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/reactome.pathway/': {'description': 'reacomte pathway id',\n",
       "  'preferred_name': 'reactome.pathway',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': '(^R-[A-Z]{3}-\\\\d+(-\\\\d+)?(\\\\.\\\\d+)?$)|(^REACT_\\\\d+(\\\\.\\\\d+)?$)',\n",
       "  'prefix': 'reactome.pathway',\n",
       "  'example': 'R-HSA-109582',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/reactome.displayname/': {'description': 'reacomte displayname id',\n",
       "  'preferred_name': 'reactome.displayname',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': None,\n",
       "  'prefix': 'reactome.displayname',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/pathways-name/': {'description': 'description for pathway',\n",
       "  'preferred_name': 'pathwaysname',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': None,\n",
       "  'prefix': 'pathwaysname',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://identifiers.org/biocarta.pathway/': {'description': 'BioCarta is a supplier and distributor of characterized reagents and assays for biopharmaceutical and academic research. It catalogs community produced online maps depicting molecular relationships from areas of active research, generating classical pathways as well as suggestions for new pathways. This collections references pathway maps.',\n",
       "  'preferred_name': 'BioCarta Pathway',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': '^([hm]\\\\_)?\\\\w+Pathway$',\n",
       "  'prefix': 'biocarta.pathway',\n",
       "  'example': 'cellcyclepathway',\n",
       "  'attribute type': 'Name'},\n",
       " 'http://identifiers.org/kegg.pathway/': {'description': 'KEGG PATHWAY is a collection of manually drawn pathway maps representing our knowledge on the molecular interaction and reaction networks.',\n",
       "  'preferred_name': 'KEGG Pathway',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': '^\\\\w{2,4}\\\\d{5}$',\n",
       "  'prefix': 'kegg.pathway',\n",
       "  'example': 'hsa00620',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/wikipathways/': {'description': 'WikiPathways is a resource providing an open and public collection of  pathway maps created and curated by the community in a Wiki like style.',\n",
       "  'preferred_name': 'WikiPathways',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': 'WP\\\\d{1,5}(\\\\_r\\\\d+)?$',\n",
       "  'prefix': 'wikipathways',\n",
       "  'example': 'WP100',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/hp/': {'description': 'The Human Phenotype Ontology (HPO) aims to provide a standardized vocabulary of phenotypic abnormalities encountered in human disease. Each term in the HPO describes a phenotypic abnormality, such as atrial septal defect. The HPO is currently being developed using the medical literature, Orphanet, DECIPHER, and OMIM.',\n",
       "  'preferred_name': 'Human Phenotype Ontology',\n",
       "  'semantic type': 'phenotype',\n",
       "  'pattern': '^HP:\\\\d{7}$',\n",
       "  'prefix': 'hp',\n",
       "  'example': 'HP:0000118',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/phenotype-name/': {'description': 'name of phenotype',\n",
       "  'preferred_name': 'phenotype-name',\n",
       "  'semantic type': 'phenotype',\n",
       "  'pattern': None,\n",
       "  'prefix': 'phenotype-name',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/phenotype-evidence/': {'description': 'phenotype evidence',\n",
       "  'preferred_name': 'phenotype-evidence',\n",
       "  'semantic type': 'phenotype',\n",
       "  'pattern': None,\n",
       "  'prefix': 'phenotype-evidence',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/phenotype-description/': {'description': 'description of phenotype',\n",
       "  'preferred_name': 'phenotype-description',\n",
       "  'semantic type': 'phenotype',\n",
       "  'pattern': None,\n",
       "  'prefix': 'phenotype-description',\n",
       "  'example': None,\n",
       "  'attribute type': 'Description'},\n",
       " 'http://biothings.io/explorer/vocab/terms/ensembl.protein/': {'description': 'Ensembl is a joint project between EMBL - EBI and the Sanger Institute to develop a software system which produces and maintains automatic annotation on selected eukaryotic genomes. This collections also references outgroup organisms',\n",
       "  'preferred_name': 'ensembl.protein',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^((ENS[FPTG]\\\\\\\\d{11}(\\\\\\\\.\\\\\\\\d+)?)|(FB\\\\\\\\w{2}\\\\\\\\d{7})|(Y[A-Z]{2}\\\\\\\\d{3}[a-zA-Z](\\\\\\\\-[A-Z])?)|([A-Z_a-z0-9]+(\\\\\\\\.)?(t)?(\\\\\\\\d+)?([a-z])?))$',\n",
       "  'prefix': 'ensembl.protein',\n",
       "  'example': 'ENSP00000243067',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/uniprot/': {'description': 'The UniProt Knowledgebase (UniProtKB) is a comprehensive resource for protein sequence and functional information with extensive cross-references to more than 120 external databases. Besides amino acid sequence and a description, it also provides taxonomic data and citation information.',\n",
       "  'preferred_name': 'UniProt Knowledgebase',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^([A-N,R-Z][0-9]([A-Z][A-Z, 0-9][A-Z, 0-9][0-9]){1,2})|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9])(\\\\.\\\\d+)?$',\n",
       "  'prefix': 'uniprot',\n",
       "  'example': 'P0DP23',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pdb/': {'description': 'The Protein Data Bank is the single worldwide archive of structural data of biological macromolecules.',\n",
       "  'preferred_name': 'Protein Data Bank',\n",
       "  'semantic type': 'gene',\n",
       "  'pattern': '^[0-9][A-Za-z0-9]{3}$',\n",
       "  'prefix': 'pdb',\n",
       "  'example': '2gc4',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pubmed/': {'description': 'PubMed is a service of the U.S. National Library of Medicine that includes citations from MEDLINE and other life science journals for biomedical articles back to the 1950s.',\n",
       "  'preferred_name': 'PubMed',\n",
       "  'semantic type': 'publication',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'pubmed',\n",
       "  'example': '16333295',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/genebank-common-name/': {'description': 'genebank common name',\n",
       "  'preferred_name': 'genebank-common-name',\n",
       "  'semantic type': 'taxonomy',\n",
       "  'pattern': None,\n",
       "  'prefix': 'genebank-common-name',\n",
       "  'example': None,\n",
       "  'attribute type': 'Name'},\n",
       " 'http://biothings.io/explorer/vocab/terms/ensembl.transcript/': {'description': 'Ensembl is a joint project between EMBL - EBI and the Sanger Institute to develop a software system which produces and maintains automatic annotation on selected eukaryotic genomes. This collections also references outgroup organisms',\n",
       "  'preferred_name': 'ensembl.transcript',\n",
       "  'semantic type': 'transcript',\n",
       "  'pattern': '^((ENS[FPTG]\\\\\\\\d{11}(\\\\\\\\.\\\\\\\\d+)?)|(FB\\\\\\\\w{2}\\\\\\\\d{7})|(Y[A-Z]{2}\\\\\\\\d{3}[a-zA-Z](\\\\\\\\-[A-Z])?)|([A-Z_a-z0-9]+(\\\\\\\\.)?(t)?(\\\\\\\\d+)?([a-z])?))$',\n",
       "  'prefix': 'ensembl.transcript',\n",
       "  'example': 'ENST00000266970',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://biothings.io/explorer/vocab/terms/ensembl.translation/': {'description': 'Ensembl is a joint project between EMBL - EBI and the Sanger Institute to develop a software system which produces and maintains automatic annotation on selected eukaryotic genomes. This collections also references outgroup organisms',\n",
       "  'preferred_name': 'ensembl.translation',\n",
       "  'semantic type': 'transcript',\n",
       "  'pattern': '^((ENS[FPTG]\\\\\\\\d{11}(\\\\\\\\.\\\\\\\\d+)?)|(FB\\\\\\\\w{2}\\\\\\\\d{7})|(Y[A-Z]{2}\\\\\\\\d{3}[a-zA-Z](\\\\\\\\-[A-Z])?)|([A-Z_a-z0-9]+(\\\\\\\\.)?(t)?(\\\\\\\\d+)?([a-z])?))$',\n",
       "  'prefix': 'ensembl.translation',\n",
       "  'example': 'ENSP00000266970',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/unigene/': {'description': 'A UniGene entry is a set of transcript sequences that appear to come from the same transcription locus (gene or expressed pseudogene), together with information on protein similarities, gene expression, cDNA clone reagents, and genomic location.',\n",
       "  'preferred_name': 'UniGene',\n",
       "  'semantic type': 'transcript',\n",
       "  'pattern': '^\\\\d+$',\n",
       "  'prefix': 'unigene',\n",
       "  'example': '4900',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/dbsnp/': {'description': 'The dbSNP database is a repository for both single base nucleotide subsitutions and short deletion and insertion polymorphisms.',\n",
       "  'preferred_name': 'dbSNP',\n",
       "  'semantic type': 'variant',\n",
       "  'pattern': '^rs\\\\d+$',\n",
       "  'prefix': 'dbsnp',\n",
       "  'example': 'rs13078881',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/pharmgkb.pathways/': {'description': 'PharmGKB Pathways are drug centric, gene based, interactive pathways which focus on candidate genes and gene groups and associated genotype and phenotype data of relevance for pharmacogenetic and pharmacogenomic studies.',\n",
       "  'preferred_name': 'pharmgkb.pathway',\n",
       "  'semantic type': 'pathway',\n",
       "  'pattern': '^PA\\\\d+$',\n",
       "  'prefix': 'pharmgkb.pathway',\n",
       "  'example': 'PA165292177',\n",
       "  'attribute type': 'ID'},\n",
       " 'http://identifiers.org/clinicaltrials/': {'description': ' \"ClinicalTrials.gov provides free access to information on clinical studies for a wide range of diseases and conditions. Studies listed in the database are conducted in 175 countries\"',\n",
       "  'preferred_name': ' clinicaltrials',\n",
       "  'semantic type': ' clinicaltrail',\n",
       "  'pattern': ' ^NCT\\\\d{8}$',\n",
       "  'prefix': ' clinicaltrial',\n",
       "  'example': ' NCT00222573',\n",
       "  'attribute type': ' ID'},\n",
       " 'http://identifiers.org/efo/': {'description': 'The Experimental Factor Ontology (EFO) provides a systematic description of many experimental variables available in EBI databases. It combines parts of several biological ontologies, such as anatomy, disease and chemical compounds. The scope of EFO is to support the annotation, analysis and visualization of data handled by the EBI Functional Genomics Team.',\n",
       "  'preferred_name': 'efo',\n",
       "  'semantic type': 'phenotype',\n",
       "  'pattern': '^\\\\d{7}$',\n",
       "  'prefix': 'efo',\n",
       "  'example': '0004859',\n",
       "  'attribute type': 'ID'}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registry.bioentity_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "nodes = []\n",
    "api_list = []\n",
    "edges = []\n",
    "output_prefixes = defaultdict(list)\n",
    "input_prefixes = defaultdict(list)\n",
    "i = 1\n",
    "node_dict = {}\n",
    "for _api in registry.api_info.keys():\n",
    "    api_dict[_api] = {'input_list': [], 'output_list': []}\n",
    "for k, v in registry.endpoint_info.items():\n",
    "    api_name = v['api']\n",
    "    if api_name not in api_list:\n",
    "        nodes.append({'group': 'api', 'id': i, 'label': api_name})\n",
    "        api_list.append(api_name)\n",
    "        api_id = i\n",
    "        i += 1\n",
    "    for _output in v['output']:\n",
    "        _output_prefix = registry.bioentity_info[_output]['prefix']\n",
    "        if _output_prefix not in node_dict:\n",
    "            node_dict[_output_prefix] = i\n",
    "            nodes.append({'group': api_name, 'id': i, 'label': _output_prefix})\n",
    "            edges.append({'from': api_id, 'label': 'hasOutput', 'to': i})\n",
    "            i += 1\n",
    "        else:\n",
    "            edges.append({'from': api_id, 'label': 'hasOutput', 'to': node_dict[_output_prefix]})\n",
    "    for _input in v['input']:\n",
    "        _input_prefix = registry.bioentity_info[_input]['prefix']\n",
    "        if _input_prefix not in node_dict:\n",
    "            node_dict[_input_prefix] = i\n",
    "            nodes.append({'group': api_name, 'id': i, 'label': _input_prefix})\n",
    "            edges.append({'from': i, 'label': 'hasInput', 'to': api_id})\n",
    "            i += 1\n",
    "        else:\n",
    "            edges.append({'from': api_id, 'label': 'hasInput', 'to': node_dict[_input_prefix]})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'group': 'api', 'id': 1, 'label': 'MyGene.info API'},\n",
       " {'group': 'MyGene.info API', 'id': 2, 'label': 'ncbigene'},\n",
       " {'group': 'MyGene.info API', 'id': 3, 'label': 'hgnc.symbol'},\n",
       " {'group': 'MyGene.info API', 'id': 4, 'label': 'refseq'},\n",
       " {'group': 'MyGene.info API', 'id': 5, 'label': 'unigene'},\n",
       " {'group': 'MyGene.info API', 'id': 6, 'label': 'uniprot'},\n",
       " {'group': 'MyGene.info API', 'id': 7, 'label': 'pdb'},\n",
       " {'group': 'MyGene.info API', 'id': 8, 'label': 'ensembl.protein'},\n",
       " {'group': 'MyGene.info API', 'id': 9, 'label': 'ensembl.transcript'},\n",
       " {'group': 'MyGene.info API', 'id': 10, 'label': 'biocarta.pathway'},\n",
       " {'group': 'MyGene.info API', 'id': 11, 'label': 'kegg.pathway'},\n",
       " {'group': 'MyGene.info API', 'id': 12, 'label': 'wikipathways'},\n",
       " {'group': 'MyGene.info API', 'id': 13, 'label': 'pharmgkb.pathway'},\n",
       " {'group': 'MyGene.info API', 'id': 14, 'label': 'reactome.pathway'},\n",
       " {'group': 'MyGene.info API', 'id': 15, 'label': 'go'},\n",
       " {'group': 'MyGene.info API', 'id': 16, 'label': 'umls.gene'},\n",
       " {'group': 'MyGene.info API', 'id': 17, 'label': 'ensembl.gene'},\n",
       " {'group': 'MyGene.info API', 'id': 18, 'label': 'omim.gene'},\n",
       " {'group': 'MyGene.info API', 'id': 19, 'label': 'hgnc'},\n",
       " {'group': 'MyGene.info API', 'id': 20, 'label': 'mirbase'},\n",
       " {'group': 'api', 'id': 21, 'label': 'MyChem.info API'},\n",
       " {'group': 'MyChem.info API', 'id': 22, 'label': 'inchikey'},\n",
       " {'group': 'MyChem.info API', 'id': 23, 'label': 'snomedct'},\n",
       " {'group': 'MyChem.info API', 'id': 24, 'label': 'drug-name'},\n",
       " {'group': 'MyChem.info API', 'id': 25, 'label': 'drugbank'},\n",
       " {'group': 'MyChem.info API', 'id': 26, 'label': 'chembl.compound'},\n",
       " {'group': 'MyChem.info API', 'id': 27, 'label': 'pubchem.compound'},\n",
       " {'group': 'MyChem.info API', 'id': 28, 'label': 'umls.compound'},\n",
       " {'group': 'MyChem.info API', 'id': 29, 'label': 'mesh.compound'},\n",
       " {'group': 'MyChem.info API', 'id': 30, 'label': 'chemical-description'},\n",
       " {'group': 'MyChem.info API', 'id': 31, 'label': 'umls.disease'},\n",
       " {'group': 'MyChem.info API', 'id': 32, 'label': 'disease-name'},\n",
       " {'group': 'MyChem.info API', 'id': 33, 'label': 'meddra'},\n",
       " {'group': 'MyChem.info API', 'id': 34, 'label': 'ndfrt'},\n",
       " {'group': 'MyChem.info API', 'id': 35, 'label': 'drug-phase'},\n",
       " {'group': 'MyChem.info API', 'id': 36, 'label': 'rxcui'},\n",
       " {'group': 'MyChem.info API', 'id': 37, 'label': 'chebi'},\n",
       " {'group': 'MyChem.info API', 'id': 38, 'label': 'iuphar.ligand'},\n",
       " {'group': 'MyChem.info API', 'id': 39, 'label': 'kegg.drug'},\n",
       " {'group': 'MyChem.info API', 'id': 40, 'label': 'kegg.compound'},\n",
       " {'group': 'MyChem.info API', 'id': 41, 'label': 'unii'},\n",
       " {'group': 'MyChem.info API', 'id': 42, 'label': 'inchi'},\n",
       " {'group': 'MyChem.info API', 'id': 43, 'label': 'pharmgkb.drug'},\n",
       " {'group': 'MyChem.info API', 'id': 44, 'label': 'dbsnp'},\n",
       " {'group': 'MyChem.info API', 'id': 45, 'label': 'pubmed'},\n",
       " {'group': 'MyChem.info API', 'id': 46, 'label': ' clinicaltrial'},\n",
       " {'group': 'MyChem.info API', 'id': 47, 'label': 'pubchem.substance'},\n",
       " {'group': 'api', 'id': 48, 'label': 'MyDisease.info API'},\n",
       " {'group': 'MyDisease.info API', 'id': 49, 'label': 'disease-description'},\n",
       " {'group': 'MyDisease.info API', 'id': 50, 'label': 'doid'},\n",
       " {'group': 'MyDisease.info API', 'id': 51, 'label': 'omim.disease'},\n",
       " {'group': 'MyDisease.info API', 'id': 52, 'label': 'mondo'},\n",
       " {'group': 'MyDisease.info API', 'id': 53, 'label': 'hp'},\n",
       " {'group': 'MyDisease.info API', 'id': 54, 'label': 'orphanet'},\n",
       " {'group': 'api', 'id': 55, 'label': 'DGIdb API'},\n",
       " {'group': 'api', 'id': 56, 'label': 'BioLink API'},\n",
       " {'group': 'BioLink API', 'id': 57, 'label': 'phenotype-name'},\n",
       " {'group': 'BioLink API', 'id': 58, 'label': 'uberon'},\n",
       " {'group': 'BioLink API', 'id': 59, 'label': 'mgi'},\n",
       " {'group': 'BioLink API', 'id': 60, 'label': 'zfin'},\n",
       " {'group': 'BioLink API', 'id': 61, 'label': 'flybase'},\n",
       " {'group': 'BioLink API', 'id': 62, 'label': 'xenbase'},\n",
       " {'group': 'BioLink API', 'id': 63, 'label': 'rgd'},\n",
       " {'group': 'BioLink API', 'id': 64, 'label': 'sgd'},\n",
       " {'group': 'BioLink API', 'id': 65, 'label': 'cl'},\n",
       " {'group': 'api', 'id': 66, 'label': 'ChEMBL API'},\n",
       " {'group': 'ChEMBL API', 'id': 67, 'label': 'mesh.disease'},\n",
       " {'group': 'ChEMBL API', 'id': 68, 'label': 'efo'},\n",
       " {'group': 'api', 'id': 69, 'label': 'Disease Ontology API'},\n",
       " {'group': 'Disease Ontology API', 'id': 70, 'label': 'ncit'},\n",
       " {'group': 'api', 'id': 71, 'label': 'Reactome API'},\n",
       " {'group': 'Reactome API', 'id': 72, 'label': 'reactome.complex'},\n",
       " {'group': 'api', 'id': 73, 'label': 'Taxonomy API'},\n",
       " {'group': 'Taxonomy API', 'id': 74, 'label': 'taxonomy'},\n",
       " {'group': 'api', 'id': 75, 'label': 'PubChem API'},\n",
       " {'group': 'PubChem API', 'id': 76, 'label': 'pubchem.bioassay'},\n",
       " {'group': 'api', 'id': 77, 'label': 'EBI Ontology Service API'},\n",
       " {'group': 'EBI Ontology Service API', 'id': 78, 'label': 'anatomy'},\n",
       " {'group': 'EBI Ontology Service API',\n",
       "  'id': 79,\n",
       "  'label': 'anatomy-description'},\n",
       " {'group': 'EBI Ontology Service API', 'id': 80, 'label': 'mesh.anatomy'},\n",
       " {'group': 'EBI Ontology Service API',\n",
       "  'id': 81,\n",
       "  'label': 'phenotype-description'},\n",
       " {'group': 'EBI Ontology Service API', 'id': 82, 'label': 'godescription'},\n",
       " {'group': 'EBI Ontology Service API', 'id': 83, 'label': 'go-term'},\n",
       " {'group': 'api', 'id': 84, 'label': 'Pharos API'},\n",
       " {'group': 'Pharos API', 'id': 85, 'label': 'gene-approved-name'},\n",
       " {'group': 'Pharos API', 'id': 86, 'label': 'pharos.disease'},\n",
       " {'group': 'Pharos API', 'id': 87, 'label': 'pharos.ligand'},\n",
       " {'group': 'Pharos API', 'id': 88, 'label': 'pharos.target'},\n",
       " {'group': 'api', 'id': 89, 'label': 'LitVar API'},\n",
       " {'group': 'api', 'id': 90, 'label': 'Ensembl API'}]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'from': 1, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 1, 'label': 'hasInput', 'to': 3},\n",
       " {'from': 4, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 5, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 6, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 7, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 8, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 9, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 10, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 11, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 12, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 13, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 14, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 15, 'label': 'hasInput', 'to': 1},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 16},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 5},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 7},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 10},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 11},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 14},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 12},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 13},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 15},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 17},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 8},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 9},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 18},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 1, 'label': 'hasOutput', 'to': 20},\n",
       " {'from': 1, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 1, 'label': 'hasInput', 'to': 17},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 22},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 6},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 22},\n",
       " {'from': 23, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 22},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 25},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 27},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 28},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 29},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 30},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 23},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 31},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 33},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 34},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 35},\n",
       " {'from': 36, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 37, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 26},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 25},\n",
       " {'from': 38, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 39, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 40, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 27},\n",
       " {'from': 41, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 42, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 24},\n",
       " {'from': 43, 'label': 'hasInput', 'to': 21},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 22},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 42},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 41},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 25},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 44},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 45},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 46},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 36},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 37},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 39},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 40},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 23},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 31},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 38},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 43},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 27},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 47},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 28},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 29},\n",
       " {'from': 21, 'label': 'hasOutput', 'to': 30},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 22},\n",
       " {'from': 21, 'label': 'hasInput', 'to': 25},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 49},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 51},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 31},\n",
       " {'from': 52, 'label': 'hasInput', 'to': 48},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 31},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 44},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 31},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 31},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 54, 'label': 'hasInput', 'to': 48},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 51},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 53},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 48, 'label': 'hasOutput', 'to': 51},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 48, 'label': 'hasInput', 'to': 3},\n",
       " {'from': 55, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 55, 'label': 'hasInput', 'to': 24},\n",
       " {'from': 55, 'label': 'hasInput', 'to': 26},\n",
       " {'from': 55, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 55, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 55, 'label': 'hasInput', 'to': 3},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 57},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 53},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 58},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 51},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 54},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 52},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 59},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 60},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 61},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 62},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 63},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 64},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 59},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 14},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 11},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 19},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 58},\n",
       " {'from': 65, 'label': 'hasInput', 'to': 56},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 58},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 53},\n",
       " {'from': 56, 'label': 'hasOutput', 'to': 52},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 56, 'label': 'hasInput', 'to': 19},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 66, 'label': 'hasInput', 'to': 24},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 66, 'label': 'hasInput', 'to': 26},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 67},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 68},\n",
       " {'from': 66, 'label': 'hasOutput', 'to': 54},\n",
       " {'from': 66, 'label': 'hasInput', 'to': 26},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 67},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 70},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 23},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 51},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 69, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 69, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 71, 'label': 'hasOutput', 'to': 72},\n",
       " {'from': 71, 'label': 'hasInput', 'to': 6},\n",
       " {'from': 71, 'label': 'hasOutput', 'to': 14},\n",
       " {'from': 71, 'label': 'hasInput', 'to': 72},\n",
       " {'from': 71, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 71, 'label': 'hasInput', 'to': 14},\n",
       " {'from': 71, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 71, 'label': 'hasInput', 'to': 6},\n",
       " {'from': 73, 'label': 'hasOutput', 'to': 74},\n",
       " {'from': 73, 'label': 'hasInput', 'to': 74},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 76},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 3},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 76},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 2},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 47},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 76},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 27},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 76},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 47},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 27},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 27},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 47},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 45},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 27},\n",
       " {'from': 75, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 75, 'label': 'hasInput', 'to': 24},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 78},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 79},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 80},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 58},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 78},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 79},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 65},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 57},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 81},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 53},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 82},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 83},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 15},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 49},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 15},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 15},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 15},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 15},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 50},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 58},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 58},\n",
       " {'from': 77, 'label': 'hasOutput', 'to': 58},\n",
       " {'from': 77, 'label': 'hasInput', 'to': 58},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 6},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 85},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 2},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 18},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 86},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 87},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 88, 'label': 'hasInput', 'to': 84},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 26},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 88},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 86},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 84, 'label': 'hasInput', 'to': 87},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 32},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 50},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 51},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 31},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 88},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 3},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 87},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 24},\n",
       " {'from': 84, 'label': 'hasInput', 'to': 86},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 87},\n",
       " {'from': 84, 'label': 'hasInput', 'to': 24},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 88},\n",
       " {'from': 84, 'label': 'hasInput', 'to': 6},\n",
       " {'from': 84, 'label': 'hasOutput', 'to': 86},\n",
       " {'from': 84, 'label': 'hasInput', 'to': 32},\n",
       " {'from': 89, 'label': 'hasOutput', 'to': 45},\n",
       " {'from': 89, 'label': 'hasInput', 'to': 44},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 44},\n",
       " {'from': 90, 'label': 'hasInput', 'to': 45},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 57},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 90, 'label': 'hasInput', 'to': 44},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 74},\n",
       " {'from': 90, 'label': 'hasInput', 'to': 74},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 8},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 68},\n",
       " {'from': 90, 'label': 'hasOutput', 'to': 53},\n",
       " {'from': 90, 'label': 'hasInput', 'to': 17}]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
